## Load Libraries

```{r}
suppressPackageStartupMessages({
  library(SoupX)
  library(Seurat)
  library(BSgenome.Nfurzeri.NCBI.UINfuzMZM1.0.custom)
  library(BSgenome.Nfurzeri.NCBI.Nfu20140520.custom)
  library(BSgenome.Nfurzeri.NCBI.NfurGRZ.RIMD1)
  library(Signac)
  library(tidyverse)
  library(scDblFinder)
})

#Load Helper Functions
source("C:/Users/Christopher He/OneDrive/UCSF/Singh Lab/diapause-sc-multiome/code/helper_functions.R")

#Set Out Directories
outs_dir <- "../results_all/"
fig_path <- "../results_all/figures/"
```

## Filepaths for CellRangerArc-aggr Outputs

use relative paths

```{r}
# all aggregated
filtered_path <- "../aggr_batch1-3/filtered_feature_bc_matrix.h5"
raw_path <- "../aggr_batch1-3/raw_feature_bc_matrix.h5"
fragments_path <- "../aggr_batch1-3/atac_fragments.tsv.gz"
annotation_path <- "../genome_2015/GCF_001465895.1_Nfu_20140520_genomic_edited.gtf"


#add sample names here in order of aggr library file
sample_names <- c("25som_1", "hb_1", "hb_2", "dia1d_1", "dia2d_1", "dia6d_1", "dia6d_2", "dia1mo_1", "dia3mo_1", "dev1d_1", "dev1d_2", "dev1d_3")
```

# SoupX Ambient RNA Removal

First, we will use SoupX to remove ambient RNA which requires filtered and raw counts

```{r}
filtered_counts <- Seurat::Read10X_h5(filtered_path)
raw_counts <- Seurat::Read10X_h5(raw_path)

filtered_gex <- filtered_counts$`Gene Expression`
raw_gex <- raw_counts$`Gene Expression`
```

SoupX runs GEX from each sample individually, so keep only GEX info and split by barcode

```{r}
tocs <- vector("list", length(sample_names))
tods <- vector("list", length(sample_names))

#each sample has a suffix after the barcode ("-1", "-2",...)
#split counts matrix by sample using suffix and rename to sample names

filtered_barcodes <- colnames(filtered_gex)
split_filtered_barcodes <- split(filtered_barcodes, sub(".*-", "", filtered_barcodes))
names(split_filtered_barcodes) <- sample_names

for(i in seq_along(sample_names)) {
   tocs[[i]] <- filtered_gex[,split_filtered_barcodes[[i]]]
   tods[[i]] <- raw_gex[,split_filtered_barcodes[[i]]]
}
```

## Full SoupX Pipeline as Function

```{r}
runSoupX <- function(tod, toc) {
  #Make soup channel object
  sc <- SoupChannel(tod, toc)
  sc$soupProfile
  #Soupx needs clusters to run
  #Do basic seurat pre-processing steps:
  soupx_so <- CreateSeuratObject(counts = toc)
  soupx_so <- NormalizeData(soupx_so, verbose = F)
  soupx_so <- FindVariableFeatures(soupx_so, verbose = F)
  soupx_so <- ScaleData(soupx_so, verbose = F)
  soupx_so <- RunPCA(soupx_so, verbose = F)
  soupx_so <- FindNeighbors(soupx_so, dims = 1:20, verbose = F) 
  soupx_so <- FindClusters(soupx_so, verbose = F)
  soupx_so <- RunUMAP(soupx_so, dims = 1:20)
  #Assign cluster identity and umap to the soupchannel object 
  meta <- cbind(soupx_so@meta.data, Embeddings(soupx_so[["umap"]]))
  sc <- setClusters(sc, setNames(meta$seurat_clusters, rownames(meta)))
  sc = setDR(sc, meta[colnames(sc$toc), c("umap_1", "umap_2")])
  #Look at rho (calculated automatically)
  sc <- autoEstCont(sc)
  #Create a new object of the adjusted counts matrix based on ambient RNA 
  out <- adjustCounts(sc)
  return(list(out = out, sc = sc))
}
```

## Run SoupX on each sample

```{r}
soup_adjusted_outs <- vector("list", length(sample_names))
sc_outs <- vector("list", length(sample_names))

for(i in seq_along(sample_names)){
  soup_outs <- runSoupX(tods[[i]], tocs[[i]])
  soup_adjusted_outs[[i]] <- soup_outs$out
  sc_outs[[i]] <- soup_outs$sc
}
```

## Look at top soup genes

```{r}
for(sc in sc_outs){
  print(head(sc$soupProfile[order(sc$soupProfile$est, decreasing = T), ], n =10))
}
# head(sc_1$soupProfile[order(sc_1$soupProfile$est, decreasing = T), ], n = 20)
# head(sc_2$soupProfile[order(sc_2$soupProfile$est, decreasing = T), ], n = 20)
# head(sc_3$soupProfile[order(sc_3$soupProfile$est, decreasing = T), ], n = 20)
# head(sc_4$soupProfile[order(sc_4$soupProfile$est, decreasing = T), ], n = 20)

# Looking at the top soup genes, we see high expression of mitochondrial genes and ribosomal genes (which is to be expected and means it is working!)
cntSoggy = rowSums(sc_1$toc > 0)
cntStrained = rowSums(out_1 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed
cntSoggy = rowSums(sc_2$toc > 0)
cntStrained = rowSums(out_2 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed
cntSoggy = rowSums(sc_3$toc > 0)
cntStrained = rowSums(out_3 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed
cntSoggy = rowSums(sc_4$toc > 0)
cntStrained = rowSums(out_4 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

```

# Create Seurat/Signac Object

re-aggregate counts from each sample

```{r}
# Combine all subsets
soupXCounts <- do.call(cbind, soup_adjusted_outs)

# Reorder the columns to match the original order
soupXCounts <- soupXCounts[, colnames(filtered_gex)]

filtered_counts$`Gene Expression` <- soupXCounts
```

## add SoupX adjusted counts to seurat object

First, import and edit annotation

```{r}
#For GRZ
gff <- rtracklayer::import(annotation_path)

gff$tx_id <- gff$transcript_id
gff$gene_name <- gff$gene_id
gff$Dbxref <- gff$db_xref
gff <- gff[,c("tx_id", "gene_name", "gene_biotype", "type", "Dbxref", "gbkey")]

#remove non-genes
gff <- gff[!is.na(gff$gene_name)]

#reformat fill in gene biotype
genes <- as.data.frame(gff[,c("gene_name", "gene_biotype")])

genes_filled <- genes %>%
  group_by(gene_name) %>%
  mutate(gene_biotype = gene_biotype[1]) %>%
  ungroup()

gff$gene_id <- gff$gene_name
gff$gene_biotype <- genes_filled$gene_biotype
```

Create Seurat object with GEX and ATAC

```{r}

so_multi <- CreateSeuratObject(
  counts = filtered_counts$`Gene Expression`,
  assay = "RNA"
)

so_multi[["ATAC"]] <- CreateChromatinAssay(
  counts = filtered_counts$Peaks,
  sep = c(":", "-"),
  fragments = fragments_path,
  annotation = gff
)
```

Add metadata columns for sample of origin and developmental trajectory

```{r}
#change orig.ident to sample identity based on barcode

so_multi$orig.ident <- sample_names[as.numeric(str_split_i(rownames(so_multi@meta.data), "-",2))]


#developmental trajectory of each sample in order of aggr library
devtraj <- c("pre-dia", "pre-dia", "pre-dia", "dia", "dia", "dia", "dia", "dia", "dia", "dev", "dev", "dev")
so_multi$devtraj <- devtraj[as.numeric(str_split_i(rownames(so_multi@meta.data), "-",2))]
so_multi$devtraj <- factor(so_multi$devtraj, levels = c("pre-dia", "dev", "dia"))


#combine dev and pre-dia into dev
diadev <- c("dev", "dev", "dev", "dia", "dia", "dia", "dia", "dia", "dia", "dev", "dev", "dev")
so_multi$diadev <- diadev[as.numeric(str_split_i(rownames(so_multi@meta.data), "-",2))]
so_multi$diadev <- factor(so_multi$diadev, levels = c("dev", "dia"))



#set levels of orig.ident
meta <- so_multi@meta.data

meta$orig.ident <- factor(meta$orig.ident, 
                          levels = sample_names)
so_multi@meta.data <- meta
```

Save Seurat Object

```{r}
saveRDS(so_multi, paste0(outs_dir, "so_postSoup.rds"))

so_multi <- readRDS(paste0(outs_dir, "so_postSoup.rds"))
```

## Run scDblFinder on GEX and ATAC

Convert to single cell experiment

```{r}
sce_multi_gex <- as.SingleCellExperiment(so_multi, assay = "RNA")
sce_multi_atac <- as.SingleCellExperiment(so_multi, assay = "ATAC")

```

scDblFinder for GEX

```{r}
sce_multi_gex <- scDblFinder(sce_multi_gex, samples = "orig.ident")

table(sce_multi_gex$scDblFinder.class)


scDblFinder_results_gex <- data.frame(sce_multi_gex@colData[,(length(sce_multi_gex@colData)-3):length(sce_multi_gex@colData)]) #last 4 cols
colnames(scDblFinder_results_gex) <- paste0(colnames(scDblFinder_results_gex), ".gex")
```

scDblFinder for ATAC

```{r}
sce_multi_atac <- scDblFinder(sce_multi_atac, aggregateFeatures=TRUE, nfeatures=25, processing="normFeatures", samples = "orig.ident")

table(sce_multi_atac$scDblFinder.class)

scDblFinder_results_atac <- data.frame(sce_multi_atac@colData[,(length(sce_multi_atac@colData)-3):length(sce_multi_atac@colData)]) #last 3 cols

colnames(scDblFinder_results_atac) <- paste0(colnames(scDblFinder_results_atac),
                                             ".atac")
```

Save results to seurat object metadata

```{r}
scDblFinder_results <- cbind(scDblFinder_results_gex, scDblFinder_results_atac )

scDblFinder_results <- scDblFinder_results[rownames(so_multi@meta.data),]

so_multi@meta.data <- cbind(so_multi@meta.data, scDblFinder_results)
```

```{r}
saveRDS(so_multi, paste0(outs_dir, "so_multi_allDoublets.rds"))

so_multi_postDF <- readRDS(paste0(outs_dir, "so_multi_allDoublets.rds"))
```

Amulet for ATAC

```{r}

# # we might also give a GRanges of repeat elements, so that these regions are excluded:
# suppressPackageStartupMessages(library(GenomicRanges))
# repeats <- GRanges("chr6", IRanges(1000,2000))
# it's better to combine these with mitochondrial and sex chromosomes

# otherChroms <- GRanges(c("NC.011814.1", "NW.026539914.1"),IRanges(1L,width=19527))
# # here since I don't know what chromosome notation you'll be using I've just put them all,
# # although this will trigger a warning when combining them:
# toExclude <- otherChroms
# # we then launch the method
# res <- amulet(fragments_path, regionsToExclude = otherChroms, fullInMemory = TRUE)
```

```{r}
so_multi_postDF <- NormalizeData(so_multi_postDF)
so_multi_postDF <- FindVariableFeatures(so_multi_postDF)
so_multi_postDF<- ScaleData(so_multi_postDF)
so_multi_postDF <- RunPCA(so_multi_postDF)
so_multi_postDF <- RunUMAP(so_multi_postDF, dims = 1:30, verbose = FALSE, reduction.name = "umap_RNA")

DimPlot(object = so_multi_postDF, reduction = "umap_RNA", group.by = "scDblFinder.class.gex")

DimPlot(object = so_multi_postDF, reduction = "umap_RNA", group.by = "scDblFinder.class.atac")
```

## Final Filtering

### calculate percent mt

```{r}
so_multi_postDF[["percent.mt"]] <- PercentageFeatureSet(so_multi_postDF, pattern = "^KEG92-", assay = 'RNA')
```

### calculate ATAC QC metrics

```{r}
DefaultAssay(so_multi_postDF) <- "ATAC"

so_multi_postDF <- NucleosomeSignal(so_multi_postDF)

saveRDS(so_multi_postDF, paste0(outs_dir, "so_atac_qc.rds"))

so_multi_postDF <- TSSEnrichment(so_multi_postDF, fast = T)

saveRDS(so_multi_postDF, paste0(outs_dir, "so_atac_qc.rds"))

# so_multi_postDF <- readRDS(paste0(outs_dir, "so_atac_qc.rds"))
```

### Plot QC Metrics

```{r}
Idents(so_multi_postDF) <- "orig.ident"

#GEX QC
DefaultAssay(so_multi_postDF) <- "RNA"
plot <- VlnPlot(
  object = so_multi_postDF,
  features = c("nFeature_RNA","nCount_RNA","percent.mt"),
  ncol = 3,
)
SaveFigure(plot, "qc_vln_rna", width = 20, height = 6)

#ATAC QC
DefaultAssay(so_multi_postDF) <- "ATAC"
plot <- VlnPlot(
  object = so_multi_postDF,
  features = c("nCount_ATAC", "TSS.enrichment", "nucleosome_signal"),
  ncol = 3,
)
SaveFigure(plot, "qc_vln_atac", width = 20, height = 6)

```

### Filtering by MAD

function for calculating MAD and calling outliers outside of n MADs

```{r}
isOutlier <- function(seurat_obj, metadata_col, n, log_values = T) {
  # Extract the metadata dataframe
  metadata <- seurat_obj@meta.data
  
  # Get the column that identifies the samples
  sample_col <- metadata$orig.ident
  
  # Initialize a logical vector to store the results
  outlier_vector <- logical(nrow(metadata))
  
  # Initialize a matrix to store the bounds for each sample
  samples <- unique(sample_col)
  bounds_matrix <- matrix(NA, nrow = length(samples), ncol = 3, 
                          dimnames = list(samples, c('lower_bound', 'upper_bound', 'median_val')))
  
  # Loop through each sample
  for (sample in unique(sample_col)) {
    # Subset the metadata for the current sample
    sample_metadata <- metadata[sample_col == sample, ]
    
    # Extract the values for the specified metadata column
    values <- sample_metadata[[metadata_col]]
    if(log_values){
      values <- log1p(values)
    }
    
    # Calculate the median and MAD
    median_val <- median(values, na.rm = TRUE)
    mad_val <- mad(values, constant = 1, na.rm = TRUE)
    
    # Determine the lower and upper bounds for outliers
    lower_bound <- median_val - n * mad_val
    upper_bound <- median_val + n * mad_val
    
    # Identify outliers
    outliers <- values < lower_bound | values > upper_bound
    
    # Update the outlier vector for the current sample
    outlier_vector[sample_col == sample] <- outliers
    
    # Store the bounds in the matrix
    bounds_matrix[sample, ] <- c(expm1(lower_bound), expm1(upper_bound), expm1(median_val))
  }
  
  metadata$outliers <- outlier_vector
  bounds_df <- data.frame(bounds_matrix)
  bounds_df$orig.ident <- rownames(bounds_df)
  
  #jitter plots with cutoffs annotated
  annotation_df <- bounds_df %>%
    mutate(lower_bound_text = paste0("Lower: ", round(lower_bound, 2)),
           upper_bound_text = paste0("Upper: ", round(upper_bound, 2)),
           y_lower = lower_bound - 0.05 * (max(metadata[[metadata_col]]) - min(metadata[[metadata_col]])), # Adjust y position for lower bound text
           y_upper = upper_bound + 0.05 * (max(metadata[[metadata_col]]) - min(metadata[[metadata_col]]))) # Adjust y position for upper bound text
  
  p <- ggplot(metadata,
              aes(x = orig.ident, y = .data[[metadata_col]], color = outliers)) +
    geom_jitter(show.legend = F) + scale_color_manual(values=c("#56B4E9", "darkgrey")) + 
    geom_text(data = annotation_df,
              aes(x = orig.ident, y = y_lower, label = lower_bound_text),
              color = "black", vjust = 1) +
    geom_text(data = annotation_df,
              aes(x = orig.ident, y = y_upper, label = upper_bound_text),
              color = "black", vjust = 0) + 
    ggtitle(paste0(metadata_col, " within ", n, " MADS"))
  plot(p)
  
  return(list("outliers" = outlier_vector, "plot" = p))
}

```

```{r}
meta <- so_multi_postDF@meta.data


ggplot(data = so_multi_postDF@meta.data,
       aes(x = orig.ident, y = nucleosome_signal)) +
  geom_jitter(color = "#56B4E9") + 
  ggtitle("nucleosome signal < 2")


meta$tss_outliers <- meta$TSS.enrichment < 1.5
ggplot(data = meta,
       aes(x = orig.ident, y = TSS.enrichment, color = tss_outliers)) + 
  geom_jitter() + 
  scale_color_manual(values=c("#56B4E9", "darkgrey")) + 
  ggtitle("TSS enrichemnt > 1.5")
```

```{r}

genes_4mads <- isOutlier(so_multi_postDF, "nFeature_RNA", 4)
genes_5mads <-isOutlier(so_multi_postDF, "nFeature_RNA", 5)
genes_6mads <-isOutlier(so_multi_postDF, "nFeature_RNA", 6)


mt_3mads <- isOutlier(so_multi_postDF, "percent.mt", 3)
mt_4mads <- isOutlier(so_multi_postDF, "percent.mt", 4)


peaks_4mads <- isOutlier(so_multi_postDF, "nFeature_ATAC", 4)
peaks_5mads <- isOutlier(so_multi_postDF, "nFeature_ATAC", 5)
peaks_6mads <- isOutlier(so_multi_postDF, "nFeature_ATAC", 6)


meta <- so_multi_postDF@meta.data

meta$filter_genes_mad4 <- genes_4mads$outliers
meta$filter_genes_mad5 <- genes_5mads$outliers
meta$filter_genes_mad6 <- genes_6mads$outliers

meta$filter_peaks_mad4 <- peaks_4mads$outliers
meta$filter_peaks_mad5 <- peaks_5mads$outliers
meta$filter_peaks_mad6 <- peaks_5mads$outliers

meta$filter_mt_mad3 <- mt_3mads$outliers
meta$filter_mt_mad4 <- mt_4mads$outliers

so_multi_postDF@meta.data <- meta
```

```{r}
library(ggpubr)

so_unfiltered <- so_multi_postDF

so_filtered_strict <- subset(
  x = so_multi_postDF,
  filter_genes_mad4 == F &
  filter_peaks_mad4 == F &
  filter_mt_mad3 == F &
  nucleosome_signal < 1 &
  TSS.enrichment > 1.5 &
  scDblFinder.class.gex == "singlet"
)



so_filtered_medium <- subset(
  x = so_multi_postDF,
  filter_genes_mad5 == F &
  filter_peaks_mad5 == F &
  filter_mt_mad3 == F &
  nucleosome_signal < 1 &
  TSS.enrichment > 1.5 &
  scDblFinder.class.gex == "singlet"
)

so_filtered_lenient <- subset(
  x = so_multi_postDF,
  filter_genes_mad6 == F &
  filter_peaks_mad6 == F &
  filter_mt_mad4 == F &
  nucleosome_signal < 1 &
  TSS.enrichment > 1.5 &
  scDblFinder.class.gex == "singlet"
)
# so_filtered <- subset(
#   x = so_multi_postDF,
#   subset = nCount_ATAC < 30000 &
#     nCount_ATAC > 1000 &
#     nucleosome_signal < 1 &
#     TSS.enrichment > 1.5 &
#     nCount_RNA > 200 &
#     nCount_RNA < 20000)
so_unfiltered
so_filtered_strict
so_filtered_medium
so_filtered_lenient


strict_plot <- ggarrange(genes_4mads$plot,peaks_4mads$plot,mt_3mads$plot, ncol = 2, nrow = 2)
SaveFigure(strict_plot, "qc_strict_filter", width = 20, height = 12)

medium_plot <- ggarrange(genes_5mads$plot,peaks_5mads$plot,mt_3mads$plot, ncol = 2, nrow = 2)
SaveFigure(medium_plot, "qc_medium_filter", width = 20, height = 12)

lenient_plot <- ggarrange(genes_6mads$plot,peaks_6mads$plot,mt_4mads$plot, ncol = 2, nrow = 2)
SaveFigure(lenient_plot, "qc_lenient_filter", width = 20, height = 12)
```

```{r}
saveRDS(so_unfiltered, paste0(outs_dir, "so_unfiltered.rds"))

saveRDS(so_filtered_strict, paste0(outs_dir, "so_filtered_strict.rds"))

saveRDS(so_filtered_medium, paste0(outs_dir, "so_filtered_medium.rds"))

saveRDS(so_filtered_lenient, paste0(outs_dir, "so_filtered_lenient.rds"))
```

## Plotting

```{r}
runSeuratRNAProcessing <- function(so){
  DefaultAssay(so) <- "RNA"
  so <- NormalizeData(so)
  so <- FindVariableFeatures(so)
  so <- ScaleData(so)
  so <- RunPCA(so)
  so <- RunUMAP(so, dims = 1:30)
  so <- FindNeighbors(so, dims = 1:30)
  so <- FindClusters(so)
  return(so)
}
```

```{r}
so_unfiltered <- runSeuratRNAProcessing(so_unfiltered)
```

```{r}
p_devtraj <- DimPlot(so_unfiltered, group.by = "devtraj", reduction = "umap",
        cols = c("#FBA91A", "#2C6E49" , "#800080"),  label = TRUE)
p_orig.ident <- DimPlot(so_unfiltered, group.by = "orig.ident", reduction = "umap",
        cols = c("#E496EF","#800080", "#800080",
                 "#00E5FF", "#A5DB6B", "#0B62AF", "#0B62AF", "#2C6E49", "#55E67E",
                 "#FBA91A", "#FBA91A", "#FBA91A"))
p_scDblFinder <- DimPlot(so_unfiltered, group.by = "scDblFinder.class.gex", reduction = "umap")

p <- ggarrange(p_devtraj, p_orig.ident, p_scDblFinder, p_clusters, ncol = 2, nrow = 2)
SaveFigure(p, "unfiltered_umaps", width = 15, height = 10)

```

```{r}
so_filtered_medium <- runSeuratRNAProcessing(so_filtered_medium)
```

```{r}
#Developmental Trajcetory
p_devtraj <- DimPlot(so_filtered_medium, group.by = "devtraj", reduction = "umap",
        cols = c("#FBA91A", "#2C6E49" , "#800080"),  label = TRUE)
#Timepoint
p_orig.ident <- DimPlot(so_filtered_medium, group.by = "orig.ident", reduction = "umap",
        cols = c("#E496EF","#800080", "#800080",
                 "#00E5FF", "#A5DB6B", "#0B62AF", "#0B62AF", "#2C6E49", "#55E67E",
                 "#FBA91A", "#FBA91A", "#FBA91A"))
p_clusters <- DimPlot(so_filtered_medium, reduction = "umap")

p <- ggarrange(p_devtraj, p_orig.ident, p_clusters, ncol = 2, nrow = 2)
SaveFigure(p, "medium_filtered_umaps", width = 15, height = 10)
```

```{r}
pseudo <- AggregateExpression(so_filtered_medium, assays = "RNA",return.seurat = T, group.by = "orig.ident")

pseudo <- NormalizeData(pseudo)
pseudo <- FindVariableFeatures(pseudo)
pseudo <- RunPCA(pseudo, npcs = 7)


p <- DimPlot(pseudo, reduction = "pca",
             cols = c("#E496EF","#800080", "#800080",
                      "#00E5FF", "#A5DB6B", "#0B62AF", "#0B62AF", "#2C6E49", "#55E67E",
                      "#FBA91A", "#FBA91A", "#FBA91A"),
             label = T, repel = T, pt.size = 4) + NoLegend()

SaveFigure(p, "psuedobulk_PCA", width = 5, height = 5)
```

<https://github.com/dpcook/code_snippets/blob/main/20210511_umap_sample_density/umap_sample_density.md>

```{r}
dat <- data.frame(so_filtered_medium@meta.data)
dat$UMAP1 <- Embeddings(so_filtered_medium, "umap")[,1]
dat$UMAP2 <- Embeddings(so_filtered_medium, "umap")[,2]

#Create a data frame that doesn't contain a "sample" column. This will allow us to facet the density layer without affecting the points
dat$Sample <- dat$orig.ident

dat_bg <- dat[,-(which(colnames(dat)=="Sample"))]

density_plot <- ggplot(dat, aes(x=UMAP1, y=UMAP2)) +
  stat_density_2d(geom="raster", aes(fill=stat(ndensity)), contour=F) + #ndensity calculates the normalized density for each sample--otherwise density would be affected by the number of cells for each sample, which is variable
  geom_point(data=dat_bg, shape=16, size=0.1, alpha=0.05, color="white") +
  scale_fill_gradientn(colours=viridisLite::mako(100), name="Density") +
  scale_x_continuous(expand=c(0,0)) +
  scale_y_continuous(expand=c(0,0)) +
  facet_wrap(~Sample, nrow=3, ncol=4) +
  theme_classic() +
  theme(strip.background = element_blank(),
        strip.text = element_text(size=12, color="black"),
        axis.text=element_blank(),
        axis.title=element_blank(),
        axis.ticks=element_blank(),
        axis.line = element_blank(),
        plot.background = element_rect(fill = "transparent", color = NA),
        legend.text=element_text(size=12, color="black"),
        legend.title=element_text(size=14, color="black"))

density_plot

ggsave(density_plot,
       filename="../results_all/figures/medium_filtered_umap_density_bysample.png",
       dpi=300,
       width=12, height=8)
```

Make UMAPs for all possible combinations of timepoints

```{r}
meta <- so_filtered_medium@meta.data

meta <- meta %>% separate_wider_delim(orig.ident, delim = "_",
                                      names = c("timepoint", "replicate"), cols_remove = F)


so_filtered_medium$timepoint <- factor(meta$timepoint, levels = c("25som", "hb", "dia1d", "dia2d",
                                                                  "dia6d", "dia1mo", "dia3mo", "dev1d"))
```

```{r}
timepoints = unique(so_filtered_medium$timepoint)
colors = c("25som" = "#E496EF", "hb" = "#800080", "dia1d" = "#00E5FF", "dia2d" = "#A5DB6B", "dia6d" = "#0B62AF", "dia1mo" = "#2C6E49", "dia3mo" = "#55E67E", "dev1d" = "#FBA91A")
for(i in 2:8){
  for(set in combn(timepoints, i, simplify = F)){
    print(set)
    so_subset <- subset(so_filtered_medium,
                        timepoint %in% c(set))
    print(nrow(so_subset@meta.data))

    so_subset <- runSeuratRNAProcessing(so_subset)
    p <- DimPlot(so_subset, group.by = "timepoint", cols = colors[set]) + ggtitle(paste(set, collapse = " + "))
    SaveFigure(p, paste0("comb_umap_", paste(set, collapse = "_")), width = 7, height = 7)
  }
}
```

```{r}
saveRDS(so_filtered_medium, paste0(outs_dir, "so_final_filtered.rds"))
```

```{r}
sessionInfo()
```
