---
title: "4_cell_type_specific_analyses"
---

## Import Libraries

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(Signac)
  library(BSgenome.Nfurzeri.NCBI.Nfu20140520.custom)
  library(tidyverse)
  library(scCustomize)
  library(ComplexHeatmap)
  library(circlize)
})

#Load Helper Functions
source("C:/Users/Christopher He/OneDrive/UCSF/Singh Lab/diapause-sc-multiome/code/helper_functions.R")

#Set Out Directories
outs_dir <- "../results_all/"
fig_dir <- "../results_all/figures/"
```

## Import Seurat Object

```{r}
so <- readRDS(paste0(outs_dir, "so_wnn.rds"))
```

```{r}
#try regressing out mitochondrial

```

# Function for performing global and cell type specific DE/DA with parameters

-   min_cell_number: minimum number of cells per group
-   timepoints_1, timepoints_2: timepoints to compare
-   downsample_to: number of cells to downsample to
-   assay: either "RNA" or "ATAC"

```{r}
#function for cell global and cell type DE (dia vs dev)
GlobalCelltypeDiaDE <- function(
    min_cell_number,
    timepoints_1 = c("dia6d", "dia1mo"),
    timepoints_2 = c("hb", "dev1d"),
    output_columns = c("pct.1", "pct.2","avg_log2FC", "p_val_adj"),
    downsample_to = F,
    assay = "RNA",
    seed = NULL
)
{
  if(!is.null(seed)){
    set.seed(seed)
  }
  
  DefaultAssay(so) <- assay

  # get list of cell types which meet min cell number
  tp1_celltypes <- (as.data.frame(table((so@meta.data %>% filter(timepoint %in% timepoints_1))$annotation_broad)) %>%
    filter(Freq > min_cell_number))$Var1
  tp2_celltypes <- (as.data.frame(table((so@meta.data %>% filter(timepoint %in% timepoints_2))$annotation_broad)) %>%
    filter(Freq > min_cell_number))$Var1

  cell_types <- intersect(tp1_celltypes, tp2_celltypes)

  out <- data.frame(
    row.names = rownames(so),
    final_symbol = translate(rownames(so))
  )
  # global timepoint1 vs timepoint2
  cells.1 <- WhichCells(so, expression = timepoint %in% timepoints_1, seed = seed)
  cells.2 <- WhichCells(so, expression = timepoint %in% timepoints_2, seed = seed)
  
  if(downsample_to != F){
      cells.1 <- sample(cells.1, downsample_to)
      cells.2 <- sample(cells.2, downsample_to)
  }
  df <- FindMarkers(
    so,
    ident.1 = cells.1,
    ident.2 = cells.2,
    logfc.threshold = -Inf,
    min.pct = -Inf
  )
  df <- df[,output_columns]
  colnames(df) <- paste0(colnames(df), "_", "global")

  #merge with full out matrix
  out <- merge(out, df, by=0)
  rownames(out) <- out[,1]
  out <- out[,-1]


  #DE for each cell type
  for(i in 1:length(cell_types)){
    cell_type <- cell_types[i]

    #get barcodes for each timepoint (group)
    cells.1 = WhichCells(
      so,
      expression =
        (timepoint %in% timepoints_1) &
        (annotation_broad == cell_type),
      seed = seed
    )
    cells.2 = WhichCells(
      so,
      expression =
        (timepoint %in% timepoints_2) &
        (annotation_broad == cell_type),
      seed = seed
    )
    if(downsample_to != F){
      cells.1 <- sample(cells.1, downsample_to)
      cells.2 <- sample(cells.2, downsample_to)
    }

    #perform DE
    df <- FindMarkers(
      so,
      ident.1 = cells.1,
      ident.2 = cells.2,
      logfc.threshold = -Inf,
      min.pct = -Inf
    )
    df <- df[,output_columns]
    colnames(df) <- paste0(colnames(df), "_", cell_type)

    #merge with full out matrix
    out <- merge(out, df, by=0)
    rownames(out) <- out[,1]
    out <- out[,-1]
  }
  return(out)
}
```

# Global and Cell Type DE

```{r}
#min cells = 200, dia groups: 6d and 1mo, dev groups: hb and 1d, no downsampling
df_200cells <- GlobalCelltypeDiaDE(min_cell_number = 200)
```

## Save to File

```{r}
write.csv(df_200cells, paste0(outs_dir, "master_genes_global_and_celltype.csv"))
```

## Iterative Downsampling for DE

```{r}

#iterative downsampling
#min_cells = 400, downsample to 300, iterate 100 times

n_iters <- 100
results_list <- vector('list', n_iters)
for(i in 1:n_iters){
  df <- GlobalCelltypeDiaDE(min_cell_number = 400, downsample_to = 300)
  df$gene <- rownames(df)
  results_list[[i]] <- df
  print(paste0("iteration ", i, " complete"))
}

results <- bind_rows(results_list)

# Identify unique cell types by extracting column names
cell_types <- unique(gsub("(avg_log2FC|p_val_adj)_", "", grep("(avg_log2FC|p_val_adj)", names(results), value = TRUE)))

# Initialize an empty list to store summary dataframes
summary_list <- list()

# Iterate over each cell type and summarize data
for (cell_type in cell_types) {
  logFC_col <- paste0("avg_log2FC_", cell_type)
  pval_col <- paste0("p_val_adj_", cell_type)
  
  summary_df <- results %>%
    group_by(gene) %>%
    summarize(
      mean_logFC = mean(!!sym(logFC_col), na.rm = TRUE),
      freq_significant = mean(!!sym(pval_col) < 0.05, na.rm = TRUE)
    ) %>%
    rename_with(~ paste0(., "_", cell_type), -gene)
  
  summary_list[[cell_type]] <- summary_df
}

# Combine all summaries into a single dataframe
final_summary <- Reduce(function(x, y) full_join(x, y, by = "gene"), summary_list)

# View the final summary
print(final_summary)
```

```{r}
cell_type_essential <- final_filtered %>% filter(rank == 11) %>% arrange(freq_significant_hatching_gland)

cell_type_essential_hg <- unlist(cell_type_essential[2:7, "gene"])


hg <- data.frame(final_filtered)
rownames(hg) <- hg$gene
hg <- hg[cell_type_essential_hg,]
hg_mat <- as.matrix(hg %>% select(contains("mean")))

# Define a function to add black borders to each cell
border_fun <- function(j, i, x, y, width, height, fill) {
  grid.rect(x = x, y = y, width = width, height = height, 
            gp = gpar(col = "black", fill = fill))
}
Heatmap(hg_mat, 
        col = colorRamp2(c(-5,0,5), c("blue","white", "red")),
        cluster_rows = F, 
        cluster_columns = F,
        show_row_names = T,
        cell_fun = border_fun,
        rect_gp = gpar(type = "none"),
        heatmap_legend_param = list(title = "log2FC"),
        column_labels = c("global", "myotome", "somite", "endothelial", "blood", "neuron", "spinal cord", "neuron prog", "mb/hb", "telen.", "epidermal", "hatching gland"),
        row_labels = translate(cell_type_essential_hg),
        width = 1,
        height = 1
        )

```

# Global and Cell Type DA

```{r}
df_200cells_ATAC <- GlobalCelltypeDiaDE(min_cell_number = 200, assay = "ATAC")
```

## Save to file

```{r}

library(JASPAR2020)

master_peaks <- read.csv(paste0(outs_dir, "master_peaks.csv"), row.names = 1)
rownames(master_peaks) <- master_peaks$Row.names
master_peaks <- master_peaks[,-1]

#consolidate motifs
motif_matrix <- master_peaks %>% select(contains("MA"))
DefaultAssay(so) <- "ATAC"
colnames(motif_matrix) <- ConvertMotifID(so, id = colnames(motif_matrix))
consolidated <- apply(motif_matrix, 1, function(row) {
  columns <- colnames(motif_matrix)[which(row)]
  return(paste(columns, collapse = ", "))
})

#remove motif matrix from master peaks
master_peaks <- master_peaks[, -grep("MA", names(master_peaks))]

#add consolidated motifs
master_peaks$motifs <- consolidated

#remove old dia/dev
master_peaks <- master_peaks[,-c(11:15)]


#merge with dia/dev cell type matrix
master_peaks <- merge(master_peaks, df_200cells_ATAC, by = "row.names")
```

```{r}
write.csv(master_peaks, paste0(outs_dir, "master_peaks_global_and_celltype.csv"))
```

## Iterative Downsampling for DA

```{r}
#iterative downsampling
#min_cells = 400, downsample to 300, iterate 100 times

n_iters <- 100
atac_iter <- vector('list', n_iters)
for(i in 1:n_iters){
  df <- GlobalCelltypeDiaDE(min_cell_number = 400, downsample_to = 300, assay = "ATAC")
  df$gene <- rownames(df)
  atac_iter[[i]] <- df
  print(paste0("iteration ", i, " complete"))
}

atac_iter <- bind_rows(atac_iter)

# Identify unique cell types by extracting column names
cell_types <- unique(gsub("(avg_log2FC|p_val_adj)_", "", grep("(avg_log2FC|p_val_adj)", names(atac_iter), value = TRUE)))

# Initialize an empty list to store summary dataframes
summary_list <- list()

# Iterate over each cell type and summarize data
for (cell_type in cell_types) {
  logFC_col <- paste0("avg_log2FC_", cell_type)
  pval_col <- paste0("p_val_adj_", cell_type)
  
  summary_df <- atac_iter %>%
    group_by(NCBI) %>%
    summarize(
      mean_logFC = mean(!!sym(logFC_col), na.rm = TRUE),
      freq_significant = mean(!!sym(pval_col) < 0.05, na.rm = TRUE)
    ) %>%
    rename_with(~ paste0(., "_", cell_type), -NCBI)
  
  summary_list[[cell_type]] <- summary_df
  print(paste0(cell_type, " done"))
}

# Combine all summaries into a single dataframe
final_atac_iter <- Reduce(function(x, y) full_join(x, y, by = "NCBI"), summary_list)

# View the final summary
head(final_atac_iter)
```

# Plotting
```{r}
#read in master sheets
master_genes <- read.csv(paste0(outs_dir, "master_genes_global_and_celltype.csv"), row.names = 1)

master_peaks <- read.csv(paste0(outs_dir, "master_peaks_global_and_celltype.csv"), row.names = 1)
rownames(master_peaks) <- master_peaks$Row.names
```

## Rank

```{r}
master_genes$rank <- apply(master_genes[,  grep("p_val_adj", names(master_genes), value = TRUE)], 1, function(row) sum(row < 0.05))
master_peaks$rank <- apply(master_peaks[,  grep("p_val_adj", names(master_peaks), value = TRUE)], 1, function(row) sum(row < 0.05))

```

## Plot global diapause genes

```{r}
global_genes <- master_genes %>% filter(
  pct.1_global > 0.01 &
  p_val_adj_global < 0.05 &
  avg_log2FC_global > 1 &
  rank > 9
) %>% arrange(desc(avg_log2FC_global))


#columns to include
columns <- unique(sapply(colnames(master_genes %>% select(contains("pct.1"))), function(x) sub("pct.1_*", "", x)))
#columns <- columns[-1]
column_cols <- colors_broad[gsub("\\.", "/",columns)]
column_cols[1] <- "#2171B5"
names(column_cols) <- columns


FC <- as.matrix(global_genes %>% select(all_of(paste0("avg_log2FC_", columns))))
pval <- as.matrix(global_genes %>% select(all_of(paste0("p_val_adj_", columns))))

FC_sig <- FC * ifelse(pval < 0.05, 1, NA)
colnames(FC_sig) <- columns
rownames(FC_sig) <- global_genes$final_symbol


ct_colors = rowAnnotation(ct = columns, col = list(ct = column_cols), show_legend=F, show_annotation_name=F)
hm <- Heatmap(
  t(FC_sig),
  col = colorRamp2(breaks = c(-5,0,0.1,5,7), colors = c("blue","white", "pink", "red", "red")),
  na_col = "grey",
  cluster_rows = F, 
  cluster_columns = F,
  show_row_names = T,
  show_column_names = T,
  rect_gp = gpar(col = "white", lwd = 0.5),
  right_annotation = ct_colors,
  heatmap_legend_param = list(title = "Fold Change in Diapause, adjusted p-val < 0.05", legend_direction = "horizontal"),
  width = unit(0.4*nrow(FC_sig), "cm"),
  height = unit(0.5*ncol(FC_sig), "cm")
)

hm <- draw(hm, heatmap_legend_side = "bottom")
SaveFigure(hm, type = "pdf", "global_gene_heatmap", height = 5, width = 25)
SaveFigure(hm, type = "png", "global_gene_heatmap", height = 5, width = 25)

```

## Plot global diapause peaks

```{r}
global_peaks <- master_peaks %>% filter(
  pct.1_global > 0.01 &
  p_val_adj_global < 0.05 &
  avg_log2FC_global > 1 &
  rank > 9
) %>% arrange(desc(avg_log2FC_global))


#columns to include
columns <- unique(sapply(colnames(master_peaks %>% select(contains("pct.1"))), function(x) sub("pct.1_*", "", x)))
#columns <- columns[-1]
column_cols <- colors_broad[gsub("\\.", "/",columns)]
column_cols[1] <- "#2171B5"
names(column_cols) <- columns


FC <- as.matrix(global_peaks %>% select(all_of(paste0("avg_log2FC_", columns))))
pval <- as.matrix(global_peaks %>% select(all_of(paste0("p_val_adj_", columns))))

FC_sig <- FC * ifelse(pval < 0.05, 1, NA)
colnames(FC_sig) <- columns
rownames(FC_sig) <- translate(master_peaks[rownames(global_peaks),"gene_name"])


ct_colors = rowAnnotation(ct = columns, col = list(ct = column_cols), show_legend=F, show_annotation_name=F)
hm <- Heatmap(
  t(FC_sig),
  col = colorRamp2(breaks = c(-5,0,0.1,5,7), colors = c("blue","white", "pink", "red", "red")),
  na_col = "grey",
  cluster_rows = F, 
  cluster_columns = F,
  show_row_names = T,
  show_column_names = F,
  rect_gp = gpar(col = "white", lwd = 0.5),
  right_annotation = ct_colors,
  heatmap_legend_param = list(title = "Fold Change in Diapause, adjusted p-val < 0.05", legend_direction = "horizontal"),
  width = unit(0.4*nrow(FC_sig), "cm"),
  height = unit(0.5*ncol(FC_sig), "cm")
)

hm <- draw(hm, heatmap_legend_side = "bottom")
SaveFigure(hm, type = "pdf", "global_peak_heatmap", height = 5, width = 25)
SaveFigure(hm, type = "png", "global_peak_heatmap", height = 5, width = 25)

```

##Plot cell type diapause genes

```{r}
dia_pct_thresh <- 0.1
p_val_adj_thresh <- 0.05
fc_thresh <- 1
rank_thresh <- 3

#rank each gene by number of cell types in which it is significantly DE
master_genes$rank <- rowSums(ifelse(master_genes %>% select(contains("p_val_adj")) < p_val_adj_thresh, 1, 0) == 1)

celltypes <- unique(sapply(colnames(master_genes %>% select(contains("pct.1"))), function(x) sub("pct.1_*", "", x)))

ctd_genes <- data.frame(avg_log2FC = numeric(), gene = character(), celltype = character())
for(celltype in celltypes[2:length(celltypes)]){
  genes <- master_genes %>% filter(
    .data[[paste0("pct.1_", celltype)]] > dia_pct_thresh &
    .data[[paste0("p_val_adj_", celltype)]] < p_val_adj_thresh &
    .data[[paste0("avg_log2FC_", celltype)]] > fc_thresh &
    rank < rank_thresh
  ) %>% arrange(desc(.data[[paste0("avg_log2FC_", celltype)]])) %>% select(paste0("avg_log2FC_", celltype))
  
  if(nrow(genes) > 0){
    genes$gene <- rownames(genes)
    rownames(genes) <- NULL
    genes$celltype <- celltype
    colnames(genes) <- colnames(ctd_genes)
    ctd_genes <- rbind(ctd_genes, genes)
  }
}
```

```{r}
# Reshape the data into long format for both avg_log2FC and p_val_adj
long_ctd <- master_genes[ctd_genes,]
long_ctd$gene <- rownames(long_ctd)

long_ctd <- long_ctd %>%
  pivot_longer(cols = starts_with("avg_log2FC"), 
               names_to = "avg_log2FC_celltype", 
               values_to = "avg_log2FC") %>%
  pivot_longer(cols = starts_with("p_val_adj"), 
               names_to = "p_val_adj_celltype", 
               values_to = "p_val_adj") %>%
  mutate(celltype = gsub("avg_log2FC_|p_val_adj_", "", avg_log2FC_celltype)) %>%
  filter(gsub("avg_log2FC_|p_val_adj_", "", avg_log2FC_celltype) == gsub("avg_log2FC_|p_val_adj_", "", p_val_adj_celltype)) %>%
  select(-avg_log2FC_celltype, -p_val_adj_celltype) # Clean up the column names

long_ctd$celltype <- factor(long_ctd$celltype, unique(long_ctd$celltype))
long_ctd$gene <- factor(long_ctd$gene, unique(long_ctd$gene))
```

```{r}
p <- ggplot(long_ctd, aes(x = celltype, y = gene)) +
  geom_point(aes(size = 1-p_val_adj, color = avg_log2FC), 
             shape = 16) +  # Use a filled circle for the dots
  scale_color_gradientn(
    colors = c("blue", "blue", "white", "red", "red"),
    breaks = c(-7,-2,0,2,7),
    values = c(0,0.4,0.5,0.6,1),
    limits = c(-7,7),
    name = "dia vs dev log2FC"
  ) +
  scale_radius(
    range = c(1,5),
    labels = c("1", "0.75", "0.5", "0.25", "0"),
    name = "adjusted p-val") +  # Size scale for the logFC, adjust range as necessary
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  # Rotate x-axis labels for clarity
  labs(title = "Cell-Type Diapause Genes",
       x = "Cell Types",
       y = "Genes") +
  theme(panel.grid = element_blank()) 

SaveFigure(p, "ctd_genes_dotplot", type = "pdf", width = 10, height = length(ctd_genes)/6)

```

```{r}
#PLOT HEATMAP

#columns to include
columns <- unique(ctd_genes$celltype)
column_cols <- colors_broad[gsub("\\.", "/",columns)]
names(column_cols) <- columns

#genes to include
ctd_genes <- ctd_genes %>%
  group_by(gene) %>%
  filter(avg_log2FC == max(avg_log2FC)) %>%
  ungroup()

ctd_FC <- as.matrix(master_genes[ctd_genes$gene,] %>% select(all_of(paste0("avg_log2FC_", columns))))
ctd_pval <- as.matrix(master_genes[ctd_genes$gene,] %>% select(all_of(paste0("p_val_adj_", columns))))

ctd_FC_sig <- ctd_FC * ifelse(ctd_pval < 0.05, 1, NA)
colnames(ctd_FC_sig) <- columns
rownames(ctd_FC_sig) <- translate(ctd_genes$gene)


ct_colors = rowAnnotation(ct = columns, col = list(ct = column_cols), show_legend=F, show_annotation_name=F)
ctg_colors = HeatmapAnnotation(ctg = ctd_genes$celltype, col = list(ctg = column_cols), show_legend = F, show_annotation_name = F, simple_anno_size = unit(0.3, "cm"))

hm <- Heatmap(
  t(ctd_FC_sig),
  col = colorRamp2(breaks = c(-5,0,0.1,5,7), colors = c("blue","white", "pink", "red", "red")),
  na_col = "grey90",
  cluster_rows = F, 
  cluster_columns = F,
  show_column_names = F,
  rect_gp = gpar(col = "white", lwd = 0.5),
  right_annotation = ct_colors,
  top_annotation = ctg_colors,
  heatmap_legend_param = list(title = "Fold Change in Diapause, adjusted p-val < 0.05", legend_direction = "horizontal"),
  width = unit(0.15*nrow(ctd_FC_sig), "cm"),
  height = unit(0.5*ncol(ctd_FC_sig), "cm")
)

hm <- draw(hm, heatmap_legend_side = "bottom")
SaveFigure(hm, type = "pdf", "ctd_gene_heatmap", height = 5, width = 15)
SaveFigure(hm, type = "png", "ctd_gene_heatmap", height = 5, width = 15)

```

## Plot cell type diapause peaks

```{r}
dia_pct_thresh <- 0.1
p_val_adj_thresh <- 0.05
fc_thresh <- 1
rank_thresh <- 3

#rank each gene by number of cell types in which it is significantly DE
master_peaks$rank <- rowSums(ifelse(master_peaks %>% select(contains("p_val_adj")) < p_val_adj_thresh, 1, 0) == 1)

celltypes <- unique(sapply(colnames(master_peaks %>% select(contains("pct.1"))), function(x) sub("pct.1_*", "", x)))

ctd_peaks <- data.frame(avg_log2FC = numeric(), peak = character(), celltype = character())
for(celltype in celltypes[2:length(celltypes)]){
  peaks <- master_peaks %>% filter(
    .data[[paste0("pct.1_", celltype)]] > dia_pct_thresh &
    .data[[paste0("p_val_adj_", celltype)]] < p_val_adj_thresh &
    .data[[paste0("avg_log2FC_", celltype)]] > fc_thresh &
    rank < rank_thresh
  ) %>% arrange(desc(.data[[paste0("avg_log2FC_", celltype)]])) %>% select(paste0("avg_log2FC_", celltype))
  
  if(nrow(peaks) > 0){
    peaks$peak <- rownames(peaks)
    rownames(peaks) <- NULL
    peaks$celltype <- celltype
    colnames(peaks) <- colnames(ctd_peaks)
    ctd_peaks <- rbind(ctd_peaks, peaks)
  }
}
```

```{r}
# Reshape the data into long format for both avg_log2FC and p_val_adj
long_ctd_peaks <- master_peaks[master_peaks$final_symbol %in% ctd_peaks,]
# long_ctd_peaks$gene <- rownames(long_ctd_peaks)

long_ctd_peaks <- long_ctd_peaks %>%
  pivot_longer(cols = starts_with("avg_log2FC"), 
               names_to = "avg_log2FC_celltype", 
               values_to = "avg_log2FC") %>%
  pivot_longer(cols = starts_with("p_val_adj"), 
               names_to = "p_val_adj_celltype", 
               values_to = "p_val_adj") %>%
  mutate(celltype = gsub("avg_log2FC_|p_val_adj_", "", avg_log2FC_celltype)) %>%
  filter(gsub("avg_log2FC_|p_val_adj_", "", avg_log2FC_celltype) == gsub("avg_log2FC_|p_val_adj_", "", p_val_adj_celltype)) %>%
  select(-avg_log2FC_celltype, -p_val_adj_celltype) # Clean up the column names


long_ctd_peaks$celltype <- factor(long_ctd_peaks$celltype, unique(long_ctd_peaks$celltype))
long_ctd_peaks$final_symbol <- factor(long_ctd_peaks$final_symbol, ctd_peaks)
```

```{r}
p <- ggplot(long_ctd_peaks, aes(x = celltype, y = final_symbol)) +
  geom_point(aes(size = 1-p_val_adj, color = avg_log2FC), 
             shape = 16) +  # Use a filled circle for the dots
  scale_color_gradientn(
    colors = c("blue", "blue", "white", "red", "red"),
    breaks = c(-7,-2,0,2,7),
    values = c(0,0.4,0.5,0.6,1),
    limits = c(-7,7),
    name = "dia vs dev log2FC"
  ) +
  scale_radius(
    range = c(1,5),
    labels = c("1", "0.75", "0.5", "0.25", "0"),
    name = "adjusted p-val") +  # Size scale for the logFC, adjust range as necessary
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  # Rotate x-axis labels for clarity
  labs(title = "Cell-Type Diapause Peaks",
       x = "Cell Types",
       y = "Genes") +
  theme(panel.grid = element_blank()) 


SaveFigure(p, "ctd_peaks_dotplot", type = "pdf", width = 10, height = length(ctd_peaks)/6)
```

```{r}
#PLOT HEATMAP

#columns to include
columns <- unique(ctd_peaks$celltype)
column_cols <- colors_broad[gsub("\\.", "/",columns)]
names(column_cols) <- columns

#peaks to include
ctd_peaks <- (ctd_peaks %>%
  group_by(peak) %>%
  filter(avg_log2FC == max(avg_log2FC)) %>%
  ungroup())


ctd_FC <- as.matrix(master_peaks[ctd_peaks$peak,] %>% select(all_of(paste0("avg_log2FC_", columns))))
ctd_pval <- as.matrix(master_peaks[ctd_peaks$peak,] %>% select(all_of(paste0("p_val_adj_", columns))))

ctd_FC_sig <- ctd_FC * ifelse(ctd_pval < 0.05, 1, NA)
colnames(ctd_FC_sig) <- columns

ct_colors = rowAnnotation(ct = columns, col = list(ct = column_cols), show_legend=F, show_annotation_name=F)
ctp_colors = HeatmapAnnotation(ctp = ctd_peaks$celltype, col = list(ctp = column_cols), show_legend = F, show_annotation_name = T, simple_anno_size = unit(0.3, "cm"))

hm <- Heatmap(
  t(ctd_FC_sig),
  col = colorRamp2(breaks = c(-5,0,0.1,5,7), colors = c("blue","white", "pink", "red", "red")),
  na_col = "grey90",
  cluster_rows = F, 
  cluster_columns = F,
  show_row_names = T,
  show_column_names = F,
  rect_gp = gpar(col = "white", lwd = 0.5),
  right_annotation = ct_colors,
  bottom_annotation = ctg_colors,
  heatmap_legend_param = list(title = "Fold Change in Diapause, adjusted p-val < 0.05", legend_direction = "horizontal"),
  width = unit(0.15*nrow(ctd_FC_sig), "cm"),
  height = unit(0.5*ncol(ctd_FC_sig), "cm")
)

hm <- draw(hm, heatmap_legend_side = "bottom")
SaveFigure(hm, type = "pdf", "ctd_peak_heatmap", height = 5, width = 25)
SaveFigure(hm, type = "png", "ctd_peak_heatmap", height = 5, width = 25)

```

## Specific Examples

```{r}
DefaultAssay(so) <- "RNA"
FeaturePlot(so, "sptssb", reduction = "umap")
VlnPlot(so, "sptssb", group.by = "annotation_broad", split.by = "devtraj" )

FeaturePlot(so, "zswim5", reduction = "umap")
VlnPlot(so, "zswim5", group.by = "annotation_broad", split.by = "devtraj" )


FeaturePlot(so, "LOC107391795", reduction = "umap")
VlnPlot(so, "LOC107391795", group.by = "annotation_broad", split.by = "devtraj" )

```

```{r}
#Function to plot cell type peaks and cell type genes
plot_ctd_peak <- function(
    so,
    peak,
    ident,
    tp1,
    tp2,
    tp1.name = "dia",
    tp1.color = "#6BAED6",
    tp2.name = "dev",
    tp2.color = "#FD8D3C",
    extend.upstream = 5000,
    extend.downstream = 5000
){
  DefaultAssay(so) <- "ATAC"
  
  so$tp <- ifelse(so$timepoint %in% tp1, tp1.name, 
                  ifelse(so$timepoint %in% tp2, tp2.name, "other"))
  so$tp <- factor(so$tp, levels = c(tp1.name, tp2.name, "other"))
  
  Idents(so) <- "annotation_broad"
  so_sub <- subset(so, idents = ident)
  l <- levels(so_sub$tp)
  so_sub$tp <- paste0(so_sub$tp, "_", ident)
  so_sub$tp <- factor(so_sub$tp, levels = paste0(l, "_", ident))

  Idents(so_sub) <- "tp"
  ct_cov <- CoveragePlot(
    object = so_sub,
    region = peak,
    annotation = F,
    peaks = F,
    links = F,
    idents = levels(so_sub$tp)[1:2],
    extend.upstream = extend.upstream,
    extend.downstream = extend.downstream,
  ) & scale_fill_manual(values = c(tp1.color, tp2.color))
  
  Idents(so) <- "tp"
  global_cov <- CoveragePlot(
    object = so,
    region = peak,
    annotation = F,
    peaks = F,
    links = F,
    idents = levels(so$tp)[1:2],
    extend.upstream = extend.upstream,
    extend.downstream = extend.downstream,
  ) & scale_fill_manual(values = c(tp1.color, tp2.color))
  
  peak_plot <- PeakPlot(
    so,
    region = peak,
    extend.upstream = extend.upstream,
    extend.downstream = extend.downstream
  )
  
  gene_plot <- AnnotationPlot(
    so,
    region = peak,
    extend.upstream = extend.upstream,
    extend.downstream = extend.downstream
  )
  
  
  p <- CombineTracks(
    plotlist = list(ct_cov, global_cov, peak_plot, gene_plot),
    heights = c(10,10,1,2)
  )
  return(p)
}
```

```{r}
p <- plot_ctd_peak(
  so,
  peak = "NC.029650.1-48333208-48334035",
  ident = "differentiating_neuron", 
  tp1 = c("dia6d", "dia1mo"),
  tp2 = c("hb", "dev1d")
)

SaveFigure(p, "ctd_peak_example", width = 6, height = 6)

p <- plot_ctd_peak(
  so,
  peak = "NC.029663.1-20428362-20429190",
  ident = "differentiating_neuron", 
  tp1 = c("dia6d", "dia1mo"),
  tp2 = c("hb", "dev1d")
)

SaveFigure(p, "ctd_peak_example_2", width = 6, height = 6)
```

```{r}
plot_ctd_gene <- function(
    so,
    gene,
    ident,
    tp1,
    tp2,
    tp1.name = "dia",
    tp1.color = "#6BAED6",
    tp2.name = "dev",
    tp2.color = "#FD8D3C"
){
  DefaultAssay(so) <- "RNA"
  so$tp <- ifelse(so$timepoint %in% tp1, tp1.name, 
                  ifelse(so$timepoint %in% tp2, tp2.name, "other"))
  so$tp <- factor(so$tp, levels = c(tp1.name, tp2.name, "other"))
  
  # Idents(so) <- "annotation_broad"
  # so_sub <- subset(so, idents = ident)
  # l <- levels(so_sub$tp)
  # so_sub$tp <- paste0(so_sub$tp, "_", ident)
  # so_sub$tp <- factor(so_sub$tp, levels = paste0(l, "_", ident))
  
  so$ct <- ifelse(so$annotation_broad == ident, ident, "all others")
  
  Idents(so) <- "tp"
  p <- VlnPlot(
    so,
    features = gene,
    group.by = "ct",
    split.by = "tp",
    idents = levels(so$tp)[1:2],
    cols = c(tp1.color, tp2.color),
    pt.size = 0)
  p <- p + geom_jitter(mapping = aes(color = split), data = p$data,
    position = position_jitterdodge(jitter.width = 0.4, dodge.width = 0.9))+
    scale_color_manual(breaks = c(tp1.name,tp2.name), values = c(tp1.color, tp2.color))
  return(p)
}
```

```{r}
plot_ctd_gene(so, "LOC107380958", "differentiating_neuron", tp1 = c("dia6d", "dia1mo"), tp2 = c("hb", "dev1d"))

for(i in 41:nrow(ctd_genes)){
  p<-plot_ctd_gene(so, ctd_genes$gene[i], ctd_genes$celltype[i], tp1 = c("dia6d", "dia1mo"), tp2 = c("hb", "dev1d"))
  SaveFigure(p, paste0("/ctd_example_genes/", i), width = 5, height = 5)
}

```

# Cell Type Level GSEA

```{r}
#read in master sheets
master_genes <- read.csv(paste0(outs_dir, "master_genes_global_and_celltype.csv"), row.names = 1)

master_peaks <- read.csv(paste0(outs_dir, "master_peaks_global_and_celltype.csv"), row.names = 1)
```

```{r}
library(msigdbr)
library(fgsea)
```

```{r}
msigdbr_species()

m_df<- msigdbr(species = "Danio rerio", category = "C5")

fgsea_sets<- m_df %>% split(x = .$gene_symbol, f = .$gs_name)

```

```{r}
dia_pct_thresh <- 0.1
p_val_adj_thresh <- 0.05

celltypes <- unique(sapply(colnames(master_peaks %>% select(contains("pct.1"))), function(x) sub("pct.1_*", "", x)))



gsea_list <- vector()

ct <- celltypes[1]
for(ct in celltypes){
  df <- master_genes
  df$ranking <- sign(df[[paste0("avg_log2FC_", ct)]])*(-log10(df[[paste0("p_val_adj_", ct)]] + .Machine$double.xmin))
  rankings <- c(df$ranking)
  names(rankings) <- translate(rownames(df), to = "Zebrafish")
  rankings <- sort(rankings, decreasing = TRUE)
  rankings <- rankings[!duplicated(names(rankings))]
  fgseaRes <- fgsea(fgsea_sets, stats = rankings)
  gsea_list[[ct]] <- fgseaRes
}
```

```{r}
saveRDS(gsea_list, paste0(outs_dir, "cell_type_gsea.rds"))
gsea_list <- readRDS(paste0(outs_dir, "cell_type_gsea.rds"))
```

```{r}
fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))


fgseaResTidy %>% 
  dplyr::select(-leadingEdge, -ES) %>% 
  arrange(padj) %>% 
  head()
```

```{r}
ggplot(fgseaResTidy %>% filter(padj < 0.008) %>% head(n= 20), aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill= NES < 7.5)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA") + 
  theme_minimal()
```

```{r}
plotEnrichment(fgsea_sets[["GOBP_FATTY_ACID_DERIVATIVE_METABOLIC_PROCESS"]],
               rankings) + labs(title="GOBP fatty acid")
```

## Cell Type motifs

```{r}
dia_pct_thresh <- 0.01
p_val_adj_thresh <- 0.05
fc_thresh <- 0.5

celltypes <- unique(sapply(colnames(master_peaks %>% select(contains("pct.1"))), function(x) sub("pct.1_*", "", x)))


ct_motifs <- data.frame(fold.enrichment = numeric(), motif.name = character(), motif = character(), celltype = character())

for(celltype in celltypes[2:length(celltypes)]){
  peaks <- master_peaks %>% filter(
    .data[[paste0("pct.1_", celltype)]] > dia_pct_thresh &
    .data[[paste0("p_val_adj_", celltype)]] < p_val_adj_thresh &
    .data[[paste0("avg_log2FC_", celltype)]] > fc_thresh
  ) %>% arrange(desc(.data[[paste0("avg_log2FC_", celltype)]])) %>% select(paste0("avg_log2FC_", celltype))
  
  enriched.motifs <- FindMotifs(
    object = so,
    features = rownames(peaks)
  )
  
  motifs <- enriched.motifs %>% filter(p.adjust < 0.05) %>% select(fold.enrichment, motif.name, motif)
  
  if(nrow(motifs) > 0){
    motifs$celltype <- celltype
    rownames(motifs) <- NULL
    ct_motifs <- rbind(ct_motifs, motifs)
  }
}

ct_motifs <- ct_motifs %>%
  select(fold.enrichment, motif.name, celltype) %>%
  pivot_wider(
    names_from = celltype,
    values_from = fold.enrichment)

ct_motifs$rank <- rowSums(!is.na(ct_motifs))-1

```

```{r}
group_global <- ct_motifs %>% filter(rank >= 4)
group_global$max_col <- "global"

group_celltype <- ct_motifs %>% filter(rank >= 1 & rank <= 3)

group_global <- group_global %>% arrange(desc(rank))



group_celltype <- group_celltype %>%
  rowwise() %>%
  mutate(max_col = names(group_celltype)[which.max(c_across(2:15))+1]) %>%
  ungroup()
group_celltype$max_col <- factor(group_celltype$max_col, levels = names(group_celltype)[2:15])
group_celltype <- group_celltype %>% arrange(max_col)

ct_motifs <- bind_rows(group_global, group_celltype) %>% select(-rank)
```

```{r}
ct_motif_mat <- as.matrix(ct_motifs %>% column_to_rownames(var = "motif.name") %>% select(-max_col))


columns <- colnames(ct_motif_mat)
column_cols <- colors_broad[gsub("\\.", "/",columns)]
names(column_cols) <- columns


motif_names <- rownames(ct_motif_mat)
motif_names[-seq(1, length(motif_names), by = 5)] <- ""
rownames(ct_motif_mat) <- motif_names

ct_colors = rowAnnotation(ct = columns, col = list(ct = column_cols), show_legend=F, show_annotation_name=F)
col_colors = HeatmapAnnotation(celltype = ct_motifs$max_col, col = list(celltype = c(global = "grey30", column_cols)), show_legend = F, show_annotation_name = F, simple_anno_size = unit(0.3, "cm"))


hm <- Heatmap(
  t(ct_motif_mat),
  col = colorRamp2(breaks = c(0,3), colors = c("white",  "red")),
  na_col = "grey90",
  cluster_rows = F, 
  cluster_columns = F,
  show_column_names = F,
  rect_gp = gpar(col = "white", lwd = 0.5),
  right_annotation = ct_colors,
  bottom_annotation = col_colors,
  heatmap_legend_param = list(title = "Fold Enrichment\n in Diapause,\n adjusted p-val < 0.05", legend_direction = "horizontal"),
)

hm <- draw(hm, heatmap_legend_side = "bottom")
SaveFigure(hm, type = "png", "ct_motif_heatmap", height = 5, width = 25)

```
```{r}
global <- MotifPlot(so, c("KLF4", "NFYB", "FOXO3", "CEBPE"))
SaveFigure(global, "global_motifs", height = 3, width = 3)

myotome <- MotifPlot(so, "POU3F4")
SaveFigure(myotome, "myotome_motifs", height = 1.5, width = 1.5)

diff_neu <- MotifPlot(so, c("CUX1", "ONECUT2"))
SaveFigure(diff_neu, "diff_neu_motifs", height = 1.5, width = 3)

neu_prog <- MotifPlot(so, c("CTCF"))
SaveFigure(neu_prog, "neu_prog_motifs", height = 1.5, width = 1.5)

peri <- MotifPlot(so, c("SOX13"))
SaveFigure(peri, "periderm_motifs", height = 1.5, width = 1.5)
```

