---
title: "4_cell_type_specific_analyses"
---

## Import Libraries

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(Signac)
  library(BSgenome.Nfurzeri.NCBI.Nfu20140520.custom)
  library(tidyverse)
  library(scCustomize)
})

#Load Helper Functions
source("C:/Users/Christopher He/OneDrive/UCSF/Singh Lab/diapause-sc-multiome/code/helper_functions.R")

#Set Out Directories
outs_dir <- "../results_all/"
fig_dir <- "../results_all/figures/"
```

## Import Seurat Object

```{r}
so <- readRDS(paste0(outs_dir, "so_wnn.rds"))
```

```{r}
#try regressing out mitochondrial

```

# Function for performing global and cell type specific DE/DA with parameters

-   min_cell_number: minimum number of cells per group
-   timepoints_1, timepoints_2: timepoints to compare
-   downsample_to: number of cells to downsample to
-   assay: either "RNA" or "ATAC"

```{r}
#function for cell global and cell type DE (dia vs dev)
GlobalCelltypeDiaDE <- function(
    min_cell_number,
    timepoints_1 = c("dia6d", "dia1mo"),
    timepoints_2 = c("hb", "dev1d"),
    output_columns = c("pct.1", "pct.2","avg_log2FC", "p_val_adj"),
    downsample_to = F,
    assay = "RNA",
    seed = NULL
)
{
  if(!is.null(seed)){
    set.seed(seed)
  }
  
  DefaultAssay(so) <- assay

  # get list of cell types which meet min cell number
  tp1_celltypes <- (as.data.frame(table((so@meta.data %>% filter(timepoint %in% timepoints_1))$annotation_broad)) %>%
    filter(Freq > min_cell_number))$Var1
  tp2_celltypes <- (as.data.frame(table((so@meta.data %>% filter(timepoint %in% timepoints_2))$annotation_broad)) %>%
    filter(Freq > min_cell_number))$Var1

  cell_types <- intersect(tp1_celltypes, tp2_celltypes)

  out <- data.frame(
    row.names = rownames(so),
    final_symbol = translate(rownames(so))
  )
  # global timepoint1 vs timepoint2
  cells.1 <- WhichCells(so, expression = timepoint %in% timepoints_1, seed = seed)
  cells.2 <- WhichCells(so, expression = timepoint %in% timepoints_2, seed = seed)
  
  if(downsample_to != F){
      cells.1 <- sample(cells.1, downsample_to)
      cells.2 <- sample(cells.2, downsample_to)
  }
  df <- FindMarkers(
    so,
    ident.1 = cells.1,
    ident.2 = cells.2,
    logfc.threshold = -Inf,
    min.pct = -Inf
  )
  df <- df[,output_columns]
  colnames(df) <- paste0(colnames(df), "_", "global")

  #merge with full out matrix
  out <- merge(out, df, by=0)
  rownames(out) <- out[,1]
  out <- out[,-1]


  #DE for each cell type
  for(i in 1:length(cell_types)){
    cell_type <- cell_types[i]

    #get barcodes for each timepoint (group)
    cells.1 = WhichCells(
      so,
      expression =
        (timepoint %in% timepoints_1) &
        (annotation_broad == cell_type),
      seed = seed
    )
    cells.2 = WhichCells(
      so,
      expression =
        (timepoint %in% timepoints_2) &
        (annotation_broad == cell_type),
      seed = seed
    )
    if(downsample_to != F){
      cells.1 <- sample(cells.1, downsample_to)
      cells.2 <- sample(cells.2, downsample_to)
    }

    #perform DE
    df <- FindMarkers(
      so,
      ident.1 = cells.1,
      ident.2 = cells.2,
      logfc.threshold = -Inf,
      min.pct = -Inf
    )
    df <- df[,output_columns]
    colnames(df) <- paste0(colnames(df), "_", cell_type)

    #merge with full out matrix
    out <- merge(out, df, by=0)
    rownames(out) <- out[,1]
    out <- out[,-1]
  }
  return(out)
}
```

# Global and Cell Type DE

```{r}
#min cells = 200, dia groups: 6d and 1mo, dev groups: hb and 1d, no downsampling
df_200cells <- GlobalCelltypeDiaDE(min_cell_number = 200)
```

## Save to File

```{r}
write.csv(df_200cells, paste0(outs_dir, "master_genes_global_and_celltype.csv"))
```

## Iterative Downsampling for DE

```{r}

#iterative downsampling
#min_cells = 400, downsample to 300, iterate 100 times

n_iters <- 100
results_list <- vector('list', n_iters)
for(i in 1:n_iters){
  df <- GlobalCelltypeDiaDE(min_cell_number = 400, downsample_to = 300)
  df$gene <- rownames(df)
  results_list[[i]] <- df
  print(paste0("iteration ", i, " complete"))
}

results <- bind_rows(results_list)

# Identify unique cell types by extracting column names
cell_types <- unique(gsub("(avg_log2FC|p_val_adj)_", "", grep("(avg_log2FC|p_val_adj)", names(results), value = TRUE)))

# Initialize an empty list to store summary dataframes
summary_list <- list()

# Iterate over each cell type and summarize data
for (cell_type in cell_types) {
  logFC_col <- paste0("avg_log2FC_", cell_type)
  pval_col <- paste0("p_val_adj_", cell_type)
  
  summary_df <- results %>%
    group_by(gene) %>%
    summarize(
      mean_logFC = mean(!!sym(logFC_col), na.rm = TRUE),
      freq_significant = mean(!!sym(pval_col) < 0.05, na.rm = TRUE)
    ) %>%
    rename_with(~ paste0(., "_", cell_type), -gene)
  
  summary_list[[cell_type]] <- summary_df
}

# Combine all summaries into a single dataframe
final_summary <- Reduce(function(x, y) full_join(x, y, by = "gene"), summary_list)

# View the final summary
print(final_summary)
```

## DE Plotting

```{r}
cols_to_check <- seq(3, ncol(final_summary), by = 2)
final_filtered <- final_summary[apply(final_summary[, cols_to_check], 1, function(row) any(row > 0.5)), ]
```

```{r}
p_freq_mat <- as.matrix(final_filtered %>% select(contains("freq_significant")))

binary_p_freq_mat <- ifelse(p_freq_mat > 0.5, 1, 0)
row_counts <- rowSums(binary_p_freq_mat == 1)


final_filtered$rank <- row_counts

global <- final_filtered %>% filter(rank > 8 & mean_logFC_global > 1)
global_mat <- as.matrix(global %>% select(contains("mean")))
rownames(global_mat) <- translate(global$gene)

library(ComplexHeatmap)
library(circlize)
Heatmap(global_mat, 
        col = colorRamp2(c(-10,0,10), c("blue","white", "red")),
        cluster_rows = T, 
        cluster_columns = T,
        show_row_names = T)
```

```{r}
ct <- final_filtered %>% filter(rank < 3 & mean_logFC_global > 1)
ct_mat <- as.matrix(ct %>% select(contains("mean")))
rownames(ct_mat) <- translate(ct$gene)

library(ComplexHeatmap)
library(circlize)
Heatmap(ct_mat, 
        col = colorRamp2(c(-10,0,10), c("blue","white", "red")),
        cluster_rows = T, 
        cluster_columns = T,
        show_row_names = T)


for(i in seq(3,25, 2)){
  genes <- (ct %>% filter(.[[i]] > 0.5))[["gene"]]
  ct_sorted <- data.frame(ct)
  rownames(ct_sorted) <- ct_sorted$gene
  ct_sorted <- ct_sorted[genes,]
  ct_mat <- as.matrix(ct_sorted %>% select(contains("mean")))
  cell_size <- unit(0.4, "cm")
  h <- Heatmap(ct_mat, 
          col = colorRamp2(c(-10,0,10), c("blue","white", "red")),
          cluster_rows = F, 
          cluster_columns = F,
          show_row_names = T,
          width = cell_size * ncol(ct_mat),
          height = cell_size * nrow(ct_mat),
          column_labels = c("global", "myotome", "somite", "endothelial", "blood", "neuron", "spinal cord", "neuron prog", "mb/hb", "telen.", "epidermal", "hatching gland"),
          heatmap_legend_param = list(title = "dia vs dev log2FC"))
  plot(h)
}



```

```{r}

PlotCellTypeGene <- function(so, gene, cell_type, annotation_column){
  so@meta.data[[cell_type]] <- ifelse(so@meta.data[[annotation_column]] == cell_type, cell_type, "all others")
  p1 <- FeaturePlot(so, features = gene, reduction = "umap")
  p2 <- VlnPlot(so, features = gene, group.by = cell_type, split.by = "devtraj", cols = c("#800080", "#6BAED6", "#FD8D3C"), pt.size = 0)
  plot(p2)
  p2 <- p2 + geom_jitter(mapping = aes(color = split), data = p2$data,
    position = position_jitterdodge(jitter.width = 0.4, dodge.width = 0.9))+
    scale_color_manual(breaks = c('pre-dia','dev', 'dia'), values = c("#800080", "#FD8D3C", "#6BAED6"))
  plot(p1)
  plot(p2)
}




PlotCellTypeGene(so, gene = "ppib", cell_type = "hatching_gland", annotation_column = "annotation_broad", pt.size = 0)
PlotCellTypeGene(so, gene = "nucb2", cell_type = "hatching_gland", annotation_column = "annotation_broad")


PlotCellTypeGene(so, gene = "LOC107385025", cell_type = "somite", annotation_column = "annotation_broad", pt.size = 0)
PlotCellTypeGene(so, gene = "stk10", cell_type = "somite", annotation_column = "annotation_broad", pt.size = 0)
PlotCellTypeGene(so, gene = "pak6", cell_type = "differentiating_neuron", annotation_column = "annotation_broad", pt.size = 0)

```

```{r}
cell_type_essential <- final_filtered %>% filter(rank == 11) %>% arrange(freq_significant_hatching_gland)

cell_type_essential_hg <- unlist(cell_type_essential[2:7, "gene"])


hg <- data.frame(final_filtered)
rownames(hg) <- hg$gene
hg <- hg[cell_type_essential_hg,]
hg_mat <- as.matrix(hg %>% select(contains("mean")))

# Define a function to add black borders to each cell
border_fun <- function(j, i, x, y, width, height, fill) {
  grid.rect(x = x, y = y, width = width, height = height, 
            gp = gpar(col = "black", fill = fill))
}
Heatmap(hg_mat, 
        col = colorRamp2(c(-5,0,5), c("blue","white", "red")),
        cluster_rows = F, 
        cluster_columns = F,
        show_row_names = T,
        cell_fun = border_fun,
        rect_gp = gpar(type = "none"),
        heatmap_legend_param = list(title = "log2FC"),
        column_labels = c("global", "myotome", "somite", "endothelial", "blood", "neuron", "spinal cord", "neuron prog", "mb/hb", "telen.", "epidermal", "hatching gland"),
        row_labels = translate(cell_type_essential_hg),
        width = 1,
        height = 1
        )

```

# Global and Cell Type DA

```{r}
df_200cells_ATAC <- GlobalCelltypeDiaDE(min_cell_number = 200, assay = "ATAC")
```

## Save to file

```{r}

library(JASPAR2020)

master_peaks <- read.csv(paste0(outs_dir, "master_peaks.csv"), row.names = 1)
rownames(master_peaks) <- master_peaks$Row.names
master_peaks <- master_peaks[,-1]

#consolidate motifs
motif_matrix <- master_peaks %>% select(contains("MA"))
DefaultAssay(so) <- "ATAC"
colnames(motif_matrix) <- ConvertMotifID(so, id = colnames(motif_matrix))
consolidated <- apply(motif_matrix, 1, function(row) {
  columns <- colnames(motif_matrix)[which(row)]
  return(paste(columns, collapse = ", "))
})

#remove motif matrix from master peaks
master_peaks <- master_peaks[, -grep("MA", names(master_peaks))]

#add consolidated motifs
master_peaks$motifs <- consolidated

#remove old dia/dev
master_peaks <- master_peaks[,-c(11:15)]


#merge with dia/dev cell type matrix
master_peaks <- merge(master_peaks, df_200cells_ATAC, by = "row.names")
```

```{r}
write.csv(master_peaks, paste0(outs_dir, "master_peaks_global_and_celltype.csv"))
```

## Iterative Downsampling for DA

```{r}
#iterative downsampling
#min_cells = 400, downsample to 300, iterate 100 times

n_iters <- 100
atac_iter <- vector('list', n_iters)
for(i in 1:n_iters){
  df <- GlobalCelltypeDiaDE(min_cell_number = 400, downsample_to = 300, assay = "ATAC")
  df$gene <- rownames(df)
  atac_iter[[i]] <- df
  print(paste0("iteration ", i, " complete"))
}
```

```{r}
atac_iter <- bind_rows(atac_iter)

# Identify unique cell types by extracting column names
cell_types <- unique(gsub("(avg_log2FC|p_val_adj)_", "", grep("(avg_log2FC|p_val_adj)", names(atac_iter), value = TRUE)))

# Initialize an empty list to store summary dataframes
summary_list <- list()

# Iterate over each cell type and summarize data
for (cell_type in cell_types) {
  logFC_col <- paste0("avg_log2FC_", cell_type)
  pval_col <- paste0("p_val_adj_", cell_type)
  
  summary_df <- atac_iter %>%
    group_by(NCBI) %>%
    summarize(
      mean_logFC = mean(!!sym(logFC_col), na.rm = TRUE),
      freq_significant = mean(!!sym(pval_col) < 0.05, na.rm = TRUE)
    ) %>%
    rename_with(~ paste0(., "_", cell_type), -NCBI)
  
  summary_list[[cell_type]] <- summary_df
  print(paste0(cell_type, " done"))
}

# Combine all summaries into a single dataframe
final_atac_iter <- Reduce(function(x, y) full_join(x, y, by = "NCBI"), summary_list)

# View the final summary
head(final_atac_iter)
```

## DA Plotting

```{r}
DefaultAssay(so) <- "ATAC"

CoveragePlot(
  object = so,
  region = "NC.029649.1-12877240-12878110",
  annotation = FALSE,
  peaks = FALSE,
  group.by = "devtraj",
  extend.upstream = 5000,
  extend.downstream = 5000
) & scale_fill_manual(values = c("#800080", "#FD8D3C", "#6BAED6"))


CoveragePlot(
  object = so,
  region = "NC.029649.1-3614145-3615014",
  annotation = T,
  peaks = T,
  group.by = "devtraj",
  extend.upstream = 5000,
  extend.downstream = 5000,
) & scale_fill_manual(values = c("#800080", "#FD8D3C", "#6BAED6"))

```

# Plotting

```{r}
#read in master sheets
master_genes <- read.csv(paste0(outs_dir, "master_genes_global_and_celltype.csv"), row.names = 1)

master_peaks <- read.csv(paste0(outs_dir, "master_peaks_global_and_celltype.csv"), row.names = 1)
```


##Plot cell type diapause genes
```{r}
dia_pct_thresh <- 0.1
p_val_adj_thresh <- 0.05
fc_thresh <- 1
rank_thresh <- 3

#rank each gene by number of cell types in which it is significantly DE
master_genes$rank <- rowSums(ifelse(master_genes %>% select(contains("p_val_adj")) < p_val_adj_thresh, 1, 0) == 1)

celltypes <- unique(sapply(colnames(master_genes %>% select(contains("pct.1"))), function(x) sub("pct.1_*", "", x)))

ctd_genes <- vector()
for(celltype in celltypes[2:length(celltypes)]){
  genes <- master_genes %>% filter(
    .data[[paste0("pct.1_", celltype)]] > dia_pct_thresh &
    .data[[paste0("p_val_adj_", celltype)]] < p_val_adj_thresh &
    .data[[paste0("avg_log2FC_", celltype)]] > fc_thresh &
    rank < rank_thresh
  )
  ctd_genes <- c(ctd_genes, rownames(genes))
}

genes <- master_genes %>% filter(
  .data[[paste0("pct.1_", "global")]] > 0.05 &
  .data[[paste0("p_val_adj_", "global")]] < p_val_adj_thresh &
  .data[[paste0("avg_log2FC_", "global")]] > fc_thresh
)

ctd_genes <- unique(ctd_genes)
```


```{r}
# Reshape the data into long format for both avg_log2FC and p_val_adj
long_ctd <- master_genes[ctd_genes,]
long_ctd$gene <- rownames(long_ctd)

long_ctd <- long_ctd %>%
  pivot_longer(cols = starts_with("avg_log2FC"), 
               names_to = "avg_log2FC_celltype", 
               values_to = "avg_log2FC") %>%
  pivot_longer(cols = starts_with("p_val_adj"), 
               names_to = "p_val_adj_celltype", 
               values_to = "p_val_adj") %>%
  mutate(celltype = gsub("avg_log2FC_|p_val_adj_", "", avg_log2FC_celltype)) %>%
  filter(gsub("avg_log2FC_|p_val_adj_", "", avg_log2FC_celltype) == gsub("avg_log2FC_|p_val_adj_", "", p_val_adj_celltype)) %>%
  select(-avg_log2FC_celltype, -p_val_adj_celltype) # Clean up the column names

long_ctd$celltype <- factor(long_ctd$celltype, unique(long_ctd$celltype))
long_ctd$gene <- factor(long_ctd$gene, unique(long_ctd$gene))
```

```{r}
p <- ggplot(long_ctd, aes(x = celltype, y = gene)) +
  geom_point(aes(size = 1-p_val_adj, color = avg_log2FC), 
             shape = 16) +  # Use a filled circle for the dots
  scale_color_gradientn(
    colors = c("blue", "blue", "white", "red", "red"),
    breaks = c(-7,-2,0,2,7),
    values = c(0,0.4,0.5,0.6,1),
    limits = c(-7,7),
    name = "dia vs dev log2FC"
  ) +
  scale_radius(
    range = c(1,5),
    labels = c("1", "0.75", "0.5", "0.25", "0"),
    name = "adjusted p-val") +  # Size scale for the logFC, adjust range as necessary
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  # Rotate x-axis labels for clarity
  labs(title = "Cell-Type Diapause Genes",
       x = "Cell Types",
       y = "Genes") +
  theme(panel.grid = element_blank()) 

SaveFigure(p, "ctd_genes_dotplot", type = "pdf", width = 10, height = length(ctd_genes)/6)

```
## Plot cell type diapause peaks

```{r}
dia_pct_thresh <- 0.1
p_val_adj_thresh <- 0.05
fc_thresh <- 1
rank_thresh <- 3

#rank each peak by number of cell types in which it is significantly DA
master_peaks$rank <- rowSums(ifelse(master_peaks %>% select(contains("p_val_adj")) < p_val_adj_thresh, 1, 0) == 1)

celltypes <- unique(sapply(colnames(master_peaks %>% select(contains("pct.1"))), function(x) sub("pct.1_*", "", x)))

ctd_peaks <- vector()
for(celltype in celltypes[2:length(celltypes)]){
  peaks <- master_peaks %>% filter(
    .data[[paste0("pct.1_", celltype)]] > dia_pct_thresh &
    .data[[paste0("p_val_adj_", celltype)]] < p_val_adj_thresh &
    .data[[paste0("avg_log2FC_", celltype)]] > fc_thresh &
    rank < rank_thresh
  )
  ctd_peaks <- c(ctd_peaks, peaks$final_symbol)
}


ctd_peaks <- unique(ctd_peaks)
```


```{r}
# Reshape the data into long format for both avg_log2FC and p_val_adj
long_ctd_peaks <- master_peaks[master_peaks$final_symbol %in% ctd_peaks,]
# long_ctd_peaks$gene <- rownames(long_ctd_peaks)

long_ctd_peaks <- long_ctd_peaks %>%
  pivot_longer(cols = starts_with("avg_log2FC"), 
               names_to = "avg_log2FC_celltype", 
               values_to = "avg_log2FC") %>%
  pivot_longer(cols = starts_with("p_val_adj"), 
               names_to = "p_val_adj_celltype", 
               values_to = "p_val_adj") %>%
  mutate(celltype = gsub("avg_log2FC_|p_val_adj_", "", avg_log2FC_celltype)) %>%
  filter(gsub("avg_log2FC_|p_val_adj_", "", avg_log2FC_celltype) == gsub("avg_log2FC_|p_val_adj_", "", p_val_adj_celltype)) %>%
  select(-avg_log2FC_celltype, -p_val_adj_celltype) # Clean up the column names


long_ctd_peaks$celltype <- factor(long_ctd_peaks$celltype, unique(long_ctd_peaks$celltype))
long_ctd_peaks$final_symbol <- factor(long_ctd_peaks$final_symbol, ctd_peaks)
```

```{r}
p <- ggplot(long_ctd_peaks, aes(x = celltype, y = final_symbol)) +
  geom_point(aes(size = 1-p_val_adj, color = avg_log2FC), 
             shape = 16) +  # Use a filled circle for the dots
  scale_color_gradientn(
    colors = c("blue", "blue", "white", "red", "red"),
    breaks = c(-7,-2,0,2,7),
    values = c(0,0.4,0.5,0.6,1),
    limits = c(-7,7),
    name = "dia vs dev log2FC"
  ) +
  scale_radius(
    range = c(1,5),
    labels = c("1", "0.75", "0.5", "0.25", "0"),
    name = "adjusted p-val") +  # Size scale for the logFC, adjust range as necessary
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +  # Rotate x-axis labels for clarity
  labs(title = "Cell-Type Diapause Peaks",
       x = "Cell Types",
       y = "Genes") +
  theme(panel.grid = element_blank()) 


SaveFigure(p, "ctd_peaks_dotplot", type = "pdf", width = 10, height = length(ctd_peaks)/6)
```
## Specific Examples
```{r}
DefaultAssay(so) <- "RNA"
FeaturePlot(so, "pclo", reduction = "umap")
VlnPlot(so, "pclo", group.by = "annotation_broad", split.by = "devtraj" )

VlnPlot(so, "LOC107380958", group.by = "annotation_broad", split.by = "devtraj" )
FeaturePlot(so, "LOC107380958", reduction = "umap")

```

```{r}

```


# Cell Type Level GSEA

```{r}
#read in master sheets
master_genes <- read.csv(paste0(outs_dir, "master_genes_global_and_celltype.csv"), row.names = 1)

master_peaks <- read.csv(paste0(outs_dir, "master_peaks_global_and_celltype.csv"), row.names = 1)
```

```{r}
library(msigdbr)
library(fgsea)
```

```{r}
msigdbr_species()

m_df<- msigdbr(species = "Danio rerio", category = "C5")
```
