---
title: "scratch"
---

```{r}
suppressPackageStartupMessages({
  library(Seurat)
  library(Signac)
  library(BSgenome.Nfurzeri.NCBI.Nfu20140520.custom)
  library(tidyverse)
  library(readxl)
  library(scCustomize)
  library(Polychrome)
})

#Load Helper Functions
source("C:/Users/Christopher He/OneDrive/UCSF/Singh Lab/diapause-sc-multiome/code/helper_functions.R")

#Set Out Directories
outs_dir <- "../results_scratch/"
fig_path <- "../results_scratch/"
```


```{r}
so <- readRDS("../results_all/so_wnn.rds")
```

## Figures for Lab Meeting 1/17

```{r}
# Pre-Filtering UMAP
so_unfiltered <- readRDS("../results_all/so_unfiltered.rds")

library(RColorBrewer)

p <- DimPlot(so_unfiltered, group.by = "devtraj", cols = c("#D94801", "#6BAED6" , "#FDD0A2"))
SaveFigure(p, "unfiltered_umap_devtraj", width = 10, height = 10)

p <- DimPlot(so_unfiltered, group.by = "orig.ident",
             cols = c("#E496EF","#800080", "#800080",
                      "#C6DBEF", "#C6DBEF", "#6BAED6", "#6BAED6", "#2171B5", "#08306B",
                      "#FD8D3C", "#FD8D3C", "#FD8D3C"))
SaveFigure(p, "unfiltered_umap_origident", width = 10, height = 10)


p <- DimPlot(so_unfiltered, group.by = "scDblFinder.class.gex",
             cols = c("grey","blue"))
SaveFigure(p, "unfiltered_umap_fdoublets", width = 10, height = 10)

```

```{r}
p <- DimPlot(so_annotated, group.by = "orig.ident",
             cols = c("#E496EF","#800080", "#800080",
                      "#C6DBEF", "#C6DBEF", "#6BAED6", "#6BAED6", "#2171B5", "#08306B",
                      "#FD8D3C", "#FD8D3C", "#FD8D3C"), reduction = "umap", pt.size = .3)
SaveFigure(p, "umap_origident", width = 10, height = 10)

p <- DimPlot(so_annotated, group.by = "devtraj",
             cols = c("#FD8D3C",
                      "#2171B5",
                      "#800080"), reduction = "umap", pt.size = .3)
SaveFigure(p, "umap_devtraj", width = 10, height = 10)

p <- FeaturePlot(so_annotated, features = "nFeature_RNA", max.cutoff = 1000, reduction = "umap")
SaveFigure(p, "umap_nFeature_RNA", width = 10, height = 10)

```

```{r}
runSeuratRNAProcessing <- function(so, resolution = 0.8){
  DefaultAssay(so) <- "RNA"
  so <- NormalizeData(so)
  so <- FindVariableFeatures(so)
  so <- ScaleData(so)
  so <- RunPCA(so)
  so <- RunUMAP(so, dims = 1:30)
  so <- FindNeighbors(so, dims = 1:30)
  so <- FindClusters(so, resolution = resolution)
  return(so)
}
so_annotated <- readRDS("../results_all/so_annotated.rds")


so_predia_dev <- subset(so_annotated, devtraj == "pre-dia" | devtraj == "dev")
so_predia_dia <- subset(so_annotated, devtraj == "pre-dia" | devtraj == "dia")
so_dia <- subset(so_annotated, devtraj == "dia")

so_predia_dev <- runSeuratRNAProcessing(so_predia_dev)
so_predia_dia <- runSeuratRNAProcessing(so_predia_dia)
so_dia <- runSeuratRNAProcessing(so_dia)


p <- DimPlot(so_predia_dev, group.by = "orig.ident",
             cols = c("#E496EF","#800080", "#800080",
                      "#FD8D3C", "#FD8D3C", "#FD8D3C"), reduction = "umap", pt.size = .3)
SaveFigure(p, "umap_predia_dev", width = 10, height = 10)

p <- DimPlot(so_predia_dia, group.by = "orig.ident",
             cols = c("#E496EF","#800080", "#800080",
                      "#C6DBEF", "#C6DBEF", "#6BAED6", "#6BAED6", "#2171B5", "#08306B"), reduction = "umap", pt.size = .3)
SaveFigure(p, "umap_predia_dia", width = 10, height = 10)

p <- DimPlot(so_dia, group.by = "orig.ident",
             cols = c("#C6DBEF", "#C6DBEF", "#6BAED6", "#6BAED6", "#2171B5", "#08306B"), reduction = "umap", pt.size = .3)
SaveFigure(p, "umap_dia", width = 10, height = 10)

```

```{r}
so_middia <- readRDS("../results_all/markers_middia/so_middia.rds")

p <- FeaturePlot(so_middia, reduction = "umap", features = "percent.mt")
SaveFigure(p, "middia_percent.mt", width = 5, height = 5)


p <- FeaturePlot(so_middia, reduction = "umap", features = "nCount_RNA", max.cutoff = 1000)
SaveFigure(p, "middia_nCount_RNA", width = 5, height = 5)


so_latedia <- readRDS("../results_all/markers_latedia/so_latedia.rds")

p <- FeaturePlot(so_latedia, reduction = "umap", features = "percent.mt")
SaveFigure(p, "latedia_percent.mt", width = 5, height = 5)


p <- FeaturePlot(so_latedia, reduction = "umap", features = "nCount_RNA", max.cutoff = 1000)
SaveFigure(p, "latedia_nCount_RNA", width = 5, height = 5)

```

```{r}
so <- readRDS("../results_all/so_annotated_atac_processed.rds")

Idents(so) <- "annotation_broad"
colors = palette36.colors(28)
colors <- sortByHue(colors)
names(colors) <- levels(Idents(so))

DefaultAssay(so) <- "RNA"
p <- DimPlot(so, reduction = "umap", cols = colors, pt.size = 0.7)
SaveFigure(p, "umap_annotated_broad", height = 10, width = 17)


DefaultAssay(so) <- "ATAC"
p <- DimPlot(so, reduction = "umap_ATAC", cols = colors, pt.size = 0.7)
SaveFigure(p, "umap_atac_annotated_broad", height = 10, width = 17)

DefaultAssay(so) <- "ATAC"
p <- DimPlot(so, reduction = "umap_ATAC", group.by = "orig.ident",
             cols = c("#E496EF","#800080", "#800080",
                      "#C6DBEF", "#C6DBEF", "#6BAED6", "#6BAED6", "#2171B5", "#08306B",
                      "#FD8D3C", "#FD8D3C", "#FD8D3C"), pt.size = 0.3)
SaveFigure(p, "umap_atac_origident", height = 10, width = 10)


DefaultAssay(so) <- "ATAC"
p <- DimPlot(so, reduction = "umap_ATAC", group.by = "devtraj",
             cols = c("#FD8D3C",
                      "#2171B5",
                      "#800080"), pt.size = .3)
SaveFigure(p, "umap_atac_devtraj", width = 10, height = 10)

DefaultAssay(so) <- "ATAC"
p <- DimPlot(so, reduction = "umap_ATAC", group.by = "ATAC_snn_res.0.8", pt.size = .3)
SaveFigure(p, "umap_atac_clusters", width = 10, height = 10)
```

```{r}
DefaultAssay(so) <- "RNA"

all_dia <- WhichCells(so, expression = devtraj == "dia")
all_dev <- WhichCells(so, expression = devtraj != "dia")

compute_DE <- function(dia, dev){
  #compute DE genes
  df <- FindMarkers(so, ident.1 = dia, ident.2 = dev)
  
  #add gene names as row, adding our annotations
  df$gene <- rownames(df)
  df$gene_name <- translate(df$gene)
  df$gene_name <- paste0(df$gene," (", df$gene_name, ")")

  #annotate with dia/dev up
  df$DE <- NA
  df$DE[df$avg_log2FC < -2 & df$p_val_adj < 0.05] <- "Dev Up"
  df$DE[df$avg_log2FC > 2 & df$p_val_adj < 0.05] <- "Dia Up"
  return(df)
}

global_diavsdev <- compute_DE(all_dia, all_dev)



p <- ggplot(global_diavsdev, aes(x = avg_log2FC, y = -log10(p_val_adj), col=DE)) + geom_point() + 
  theme_minimal() + xlim(-15, 15) +
  geom_vline(xintercept=c(-2, 2), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red") + 
  scale_color_manual(values=c("red", "blue", "black"))

SaveFigure(p, "volcano_RNA", width = 4, height = 4)

```

```{r}
DefaultAssay(so) <- "ATAC"

all_dia <- WhichCells(so, expression = devtraj == "dia")
all_dev <- WhichCells(so, expression = devtraj != "dia")

global_diavsdev_atac <- FindMarkers(so, ident.1 = all_dia, ident.2 = all_dev)

global_diavsdev_atac$DE <- NA
global_diavsdev_atac$DE[global_diavsdev_atac$avg_log2FC < -1 & global_diavsdev_atac$p_val_adj < 0.05] <- "Dev Up"
global_diavsdev_atac$DE[global_diavsdev_atac$avg_log2FC > 1 & global_diavsdev_atac$p_val_adj < 0.05] <- "Dia Up"


p <- ggplot(global_diavsdev_atac, aes(x = avg_log2FC, y = -log10(p_val_adj), col=DE)) + geom_point() + 
  theme_minimal() + xlim(-15, 15) +
  geom_vline(xintercept=c(-1, 1), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red") + 
  scale_color_manual(values=c("red", "blue", "black"))

SaveFigure(p, "volcano_ATAC", width = 4, height = 4)
```

```{r}
top.da.peak <- rownames(global_diavsdev_atac %>% filter(DE == "Dia Up"))

enriched.motifs <- FindMotifs(
  object = so,
  features = top.da.peak
)

p <- MotifPlot(
  object = so,
  motifs = head(rownames(enriched.motifs))
)

SaveFigure(p, "motifs_dia", width = 6, height = 4)

```

```{r}
DefaultAssay(so) <- "RNA"

pseudo <- AggregateExpression(so, assays = "RNA",return.seurat = T, group.by = "orig.ident")

pseudo <- NormalizeData(pseudo)
pseudo <- FindVariableFeatures(pseudo)
pseudo <- RunPCA(pseudo, npcs = 7)


p <- DimPlot(pseudo, reduction = "pca",
             cols = c("#E496EF","#800080", "#800080",
                      "#C6DBEF", "#C6DBEF", "#6BAED6", "#6BAED6", "#2171B5", "#08306B",
                      "#FD8D3C", "#FD8D3C", "#FD8D3C"),
             label = T, repel = T, pt.size = 4) + NoLegend()

SaveFigure(p, "pseudobulk_pca", width = 5, height = 5)
```

```{r}
DefaultAssay(so) <- "RNA"
so$devtraj <- factor(so$devtraj, levels = c("pre-dia", "dev", "dia"))


so$hatch <- ifelse(so$annotation_broad == "hatching_gland", "hatching_gland", "all others")


p <- FeaturePlot(so, reduction = "umap", features = "ppib") + NoLegend() + NoAxes()
SaveFigure(p, "ppib_feature", width = 5, height = 5)

p <- VlnPlot(so, features = "ppib", group.by = "hatch", split.by = "devtraj",
        cols = c("#E496EF", "#FD8D3C", "#2171B5"), pt.size = 0)
SaveFigure(p, "ppib_vln", width = 5, height = 5)



so$neuron <- ifelse(so$annotation_broad == "neuron", "neuron", "all others")
p <- FeaturePlot(so, reduction = "umap", features = "LOC107372894") + NoLegend() + NoAxes()
SaveFigure(p, "LOC107372894_feature", width = 5, height = 5)

p <- VlnPlot(so, features = "LOC107372894", group.by = "neuron", split.by = "devtraj",
        cols = c("#E496EF", "#FD8D3C", "#2171B5"),y.max = 2, pt.size = 0)
SaveFigure(p, "LOC107372894_vln", width = 5, height = 5)

```

```{r}
bulk <- read_excel("C:\\Users\\Christopher He\\OneDrive\\UCSF\\Singh Lab\\diapause-sc-multiome\\bulk\\1Diapause_RNA-seq_with_human_symbol_CK.xlsx")

rownames(bulk) <- bulk$genes

hm <- as.matrix(bulk[c("LOC107373895", top_dia$gene), c(21, 22, 23, 24, 25)])
rownames(hm) <- c("LOC107373895", top_dia$gene)


row_normalize <- function(m) {
  t(apply(m, 1, function(row) {
    (row - mean(row)) / sd(row)
  }))
}

# Apply the row normalization function to the matrix
hm <- row_normalize(hm)

# Print the normalized matrix

ht <- Heatmap(hm,
        cluster_columns = F,
        cluster_rows = F,
        row_names_gp = gpar(fontsize = 5)
)

SaveFigure(ht, "figures/heatmap_diaup_bulk", height = 10, width = 10)

```

cell type composition changes

```{r}
# Extract metadata
metadata <- so@meta.data

# Calculate cell counts per timepoint and cell type
cell_counts <- metadata %>%
  group_by(timepoint_group, annotation_broad) %>%
  summarize(count = n()) %>%
  ungroup()

# Calculate total cells per timepoint
total_cells <- cell_counts %>%
  group_by(timepoint_group) %>%
  summarize(total = sum(count))

# Merge to get percentages
cell_percentages <- cell_counts %>%
  left_join(total_cells, by = "timepoint_group") %>%
  mutate(percentage = (count / total) * 100)


# Plot percentages
percentage_plot <- ggplot(cell_percentages, aes(x = timepoint_group, y = percentage, fill = annotation_broad)) +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = colors) +
  labs(x = "Timepoint Group", y = "Percentage", fill = "Cell Type") +
  theme_minimal() +
  ggtitle("Cell Type Composition Over Time (Percentage)")

# Print the plot
SaveFigure(percentage_plot, "figures/cell_type_composition_percentage", height = 10, width = 10)

# Plot total counts
count_plot <- ggplot(cell_counts, aes(x = timepoint_group, y = count, fill = annotation_broad)) +
  geom_bar(stat = "identity", position = "stack") +
  scale_fill_manual(values = colors) +
  labs(x = "Timepoint Group", y = "Cell Count", fill = "Cell Type") +
  theme_minimal() +
  ggtitle("Cell Type Composition Over Time (Total Cell Count)")

# Print the plot
SaveFigure(count_plot, "figures/cell_type_composition_total", height = 10, width = 10)

```

```{r}
VlnPlot(so, features = "reps2", group.by = "timepoint_group")
FeaturePlot(so, features = "reps2", reduction = "umap")

VlnPlot(so, features = "sepp1", group.by = "timepoint_group")
FeaturePlot(so, features = "sepp1", reduction = "umap")

```

```{r}
annotations <- so_annotated@meta.data[, c("annotation_grouped", "annotation_merged", "annotation_broad")]


write.csv(annotations, paste0(outs_dir, "annotations.csv"))
```

```{r}
DefaultAssay(so_annotated) <- "RNA"

FeaturePlot(so_annotated, "hnf1b")
```
# HNF4a Analyses
```{r}
DefaultAssay(so) <- "RNA"

#HNF genes:
hnf_genes <- c("hnf1a", "hnf4g", "LOC107381564", "hnf4a", "hnf1b")

translate(hnf_genes)


#Jaspar Motif:
hnf_motifs <- c("MA0114.4", "MA0484.2", "MA1494.1", "MA0046.2", "MA0153.2")
hnf_motif_names <- c("HNF4A", "HNF4G", "HNF4A(var.2)", "HNF1A", "HNF1B")

p <- FeaturePlot(so, features = hnf_genes, reduction = "umap") + patchwork::plot_annotation(caption = paste(translate(hnf_genes), collapse = ', '))
SaveFigure(p, "hnf_feature", height = 10, width = 10)


DefaultAssay(so) <- "chromvar"
FeaturePlot(so, hnf_motifs)


p <- VlnPlot(so, hnf_genes, split.by = "timepoint_group", assay = "RNA")
p <- VlnPlot(so, hnf_genes, split.by = "timepoint_group", assay = "gene_activity")
p <- VlnPlot(so, hnf_motifs, split.by = "timepoint_group", assay = "chromvar")



```
what hnf peaks were opened in diapause?
```{r}
master_peaks <- read.csv("../results_all/master_peaks.csv", row.names = 1)

HNF4A_peaks_dia_up <- master_peaks[master_peaks$MA0114.4 == T,] %>% filter(p_val_adj < 0.05 & avg_log2FC > 0)


annotation_counts <- as.data.frame(table(master_peaks$peak_annotation))
colnames(annotation_counts) <- c("annotation", "count")

# Create the stacked bar chart
ggplot(annotation_counts, aes(x = "", y = count, fill = annotation)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = count), position = position_stack(vjust = 0.5), hjust = 0.5) +
  labs(x = "Annotations", y = "Count", fill = "Annotation Type") +
  ggtitle("Counts of Annotations in master_peaks") +
  theme_minimal() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())


annotation_counts <- as.data.frame(table(HNF4A_peaks_dia_up$peak_annotation))
colnames(annotation_counts) <- c("annotation", "count")

# Create the stacked bar chart
ggplot(annotation_counts, aes(x = "", y = count, fill = annotation)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = count), position = position_stack(vjust = 0.5), hjust = 0.5) +
  labs(x = "Annotations", y = "Count", fill = "Annotation Type") +
  ggtitle("Counts of Annotations in master_peaks") +
  theme_minimal() +
  theme(axis.text.x = element_blank(), axis.ticks.x = element_blank())


write.csv(translate(unique(HNF4A_peaks_dia_up$gene_name), to = "Human"), paste0(outs_dir, "HNF4A_peaks_dia_up_genes.csv"))

```

GO for HNF4a target genes
```{r}
go_data <- read.csv(paste0(outs_dir, "hnf4a_GO_mf.txt"), header = TRUE, sep = "\t")

# Preprocess data if necessary (e.g., filter for significant terms)
significant_go_data <- go_data %>%
  filter(upload_1..FDR. < 0.05) %>%
  arrange(desc(upload_1..fold.Enrichment.))

top_go_terms <- head(significant_go_data, 25)

# Create the plot
p <- ggplot(top_go_terms, aes(x = reorder(GO.molecular.function.complete, upload_1..FDR.), 
                              y = upload_1..fold.Enrichment., 
                              fill = upload_1..FDR.)) +
  geom_bar(stat = "identity") +
  scale_fill_gradient(low = "red", high = "blue") +
  coord_flip() +
  labs(title = "Top Molecular Function Terms",
       x = "GO Term",
       y = "Fold Enrichment") +
  theme_minimal()

# Save the plot
ggsave(paste0(outs_dir, "HNF4a_GOmf_barplot.png"), p, width = 15, height = 7)

go_data <- read.csv(paste0(outs_dir, "hnf4a_GO_bp.txt"), header = TRUE, sep = "\t")

# Preprocess data if necessary (e.g., filter for significant terms)
significant_go_data <- go_data %>%
  filter(upload_1..FDR. < 0.05) %>%
  arrange(desc(upload_1..fold.Enrichment.))

# Bar plot of top 10 GO terms by fold enrichment
top_go_terms <- head(significant_go_data, 25)

# Create the plot
p <- ggplot(top_go_terms, aes(x = reorder(GO.biological.process.complete, upload_1..FDR.), 
                              y = upload_1..fold.Enrichment., 
                              fill = upload_1..FDR.)) +
  geom_bar(stat = "identity") +
  scale_fill_gradient(low = "red", high = "blue") +
  coord_flip() +
  labs(title = "Top Biological Process Terms",
       x = "GO Term",
       y = "Fold Enrichment") +
  theme_minimal()

# Save the plot
ggsave(paste0(outs_dir, "HNF4a_GObp_barplot.png"), p, width = 12, height = 7)

```


```{r}
df <- data.frame(table(HNF4A_peaks_dia_up$peak_annotation))
df$group <- "HNF4a"
df2 <- data.frame(table(master_peaks$peak_annotation))
df2$group <- "all"

df <- rbind(df, df2)

ggplot(df, aes(fill=Var1, y=Freq, x=group)) + 
  geom_bar(position="fill", stat="identity")
  
  
df <- df %>%
  group_by(group) %>%
  mutate(Percent = Freq / sum(Freq) * 100, Label = paste0(round(Percent, 1), "% (", Freq, ")"))

# Create the stacked percent bar graph
ggplot(df, aes(fill = Var1, y = Percent, x = group)) +
  geom_bar(position = "fill", stat = "identity") +
  geom_text(aes(label = Label),
            position = position_fill(vjust = 0.5),
            color = "white") +
  scale_y_continuous(labels = scales::percent_format()) +
  labs(y = "Percentage", x = "Group",
       title = "Stacked Percent Bar Graph with Annotations")
```

ATAC annotations
```{r}
DefaultAssay(so) <- "ATAC"


DimPlot(object = so, reduction = "umap_ATAC", group.by = "ATAC_snn_res.0.8", label = TRUE) + NoLegend()

atac_markers <- FindAllMarkers(so, assay = "gene_activity", group.by = "ATAC_snn_res.0.8", only.pos = T)

```
```{r}
atac_markers$gene_name <- translate(atac_markers$gene)
atac_markers$gene_name <- paste0(atac_markers$gene,
                                  " (", atac_markers$gene_name, ")")

top_atac_markers <- atac_markers %>%
  filter(avg_log2FC > 0.5) %>%
  group_by(cluster) %>%
  arrange(p_val_adj) %>%
  slice_head(n = 20) %>%
  mutate(rank = row_number()) %>%
  select(cluster, rank, gene_name) %>%
  pivot_wider(names_from = cluster, values_from = gene_name, names_prefix = "Cluster_", values_fill = list(gene_name = "")) %>%
  arrange(rank) %>%
  select(-rank)

write.csv(top_atac_markers,
          paste0(outs_dir, "top_atac_markers.csv"))

```
```{r}
DefaultAssay(so) <- 'gene_activity'
FeaturePlot(so, features = c("LOC107376706", "LOC107373085", "LOC107382309"), reduction = "umap_ATAC")

FeaturePlot(so, features = c("clec3b"), reduction = "umap_ATAC")

```

```{r}
Idents(so) <- "ATAC_snn_res.0.8"

cluster.ids <- c(
  'brain_1',
  'spinal_cord_1',
  'neuron_1',
  'midbrain_hindbrain',
  'myotome/muscle_1',
  'neuron_progenitor_1',
  'somite_1',
  'somite_2',
  'telencephalon/optic_cup_1',
  'neuron_progenitor_2',
  'epidermal',
  'spinal_cord_2',
  'telencephalon/optic_cup_2',
  'spinal_cord_3',
  'telencephalon/optic_cup_3',
  'neuron_progenitor_3',
  'spinal_cord_4',
  'unknown',
  'head_kidney',
  'hatching_gland',
  'midbrain_hindbrain_pax2/5_en1/2',
  'brain_2',
  'periderm',
  'endothelial',
  'primitive_gut',
  'floor_plate',
  'spinal_cord_5',
  'neural_crest',
  'blood',
  'otic_vesicle',
  'neuron_dscam',
  'neuron_neurod1',
  'telencephalon/optic_cup_4',
  'myotome/muscle_2'
)

names(cluster.ids) <- levels(so)
so <- RenameIdents(so, cluster.ids)

so$annotation_ATAC <- Idents(so)


p <- DimPlot(so, label = T, repel = T, reduction = "umap_ATAC") + NoLegend()
SaveFigure(p, path = "umap_annotated_ATAC",
           height = 10, width = 10)
```

```{r}


so_latedia <- readRDS(paste0(outs_dir, "markers_latedia/so_latedia.rds"))

DimPlot(so_latedia, reduction = "umap", label = T)

VlnPlot(so_latedia, "lcp1")
FeaturePlot(so_latedia, "lcp1", reduction = "umap")

table(Idents(so_latedia))

```
# Parse candidates from Rui


```{r}
# Read in List of Candidates
genes <- read.csv(paste0(outs_dir, "parse_candidates_rui_20250318.csv"), row.names = 1)
genes <- genes %>% filter(Count_in_intersects > 1)

#read in master sheet
master_genes <- read.csv("../results_all/master_genes_global_and_celltype.csv", row.names = 1)
filtered_master <- master_genes[genes$Gene,]  %>% arrange(-avg_log2FC_global)

genes_FC <- as.matrix(filtered_master %>% select(contains("FC")))
genes_p_val <- as.matrix(filtered_master %>% select(contains("p_val_adj")))

FC_sig <- genes_FC * ifelse(genes_p_val < 0.05, 1, NA)


colnames(FC_sig) <- lapply(colnames(genes_FC), function(x) substring(x, 12))
rownames(FC_sig) <- translate(rownames(genes_FC))

library(circlize)
library(ComplexHeatmap)

Heatmap(
  t(FC_sig),
  col = colorRamp2(breaks = c(-5,0,5), colors = c("blue","white", "red")),
  na_col = "grey",
  cluster_rows = F, 
  cluster_columns = F,
  show_row_names = T,
  row_names_side = "left",
  show_column_names = T,
  rect_gp = gpar(col = "white", lwd = 0.5),
  heatmap_legend_param = list(title = "Log2 Fold Change in Diapause", legend_direction = "horizontal"),
)

write.csv(head(FC_sig, 20), paste0(outs_dir, "diapause_markers_parse_and_10x.csv"))
```
# Expression Markers
```{r}
DefaultAssay(so) <- "RNA"

markers <- FindAllMarkers(so, group.by = "annotation_broad")
#markers_roc <- FindAllMarkers(so, group.by = "annotation_broad", test.use = "roc")
```


```{r}
ct_markers <- markers %>%
  filter(pct.1 > 0.2 & avg_log2FC > 2 & p_val_adj < 0.005 & cluster != "hibernating_unknown") %>%
  group_by(gene) %>%
  mutate(row_count = n()) %>%
  filter(row_count < 3) %>%
  arrange(desc(pct.1)) %>%
  slice_max(pct.1, n = 1) %>%
  ungroup() %>%
  group_by(cluster) %>%
  slice_max(pct.1, n = 3)

ct_marker_names <- translate(ct_markers$gene)

DotPlot(so, features = ct_markers$gene, group.by = "annotation_broad") + RotatedAxis() + scale_x_discrete(labels=ct_marker_names)

# SaveFigure(p, "1C_marker_dotplot", type = "png", width = 15, height = 8, res = 300)
```

```{r}
#top 5 per cell type
filtered_markers <- markers %>%
  filter(pct.1 > 0.2 & avg_log2FC > 2 & p_val_adj < 0.005 & cluster != "hibernating_unknown") %>%
  group_by(gene) %>%
  mutate(row_count = n()) %>%
  ungroup() %>% mutate(final_symbol = translate(gene)) %>%
  select(gene, final_symbol, avg_log2FC, pct.1, pct.2, cluster, row_count)

ct_markers <- filtered_markers %>%
  group_by(cluster) %>%
  slice_max(pct.1, n = 5) %>%
  ungroup()

write.csv(ct_markers, row.names = F, paste0(outs_dir, "ct_markers.csv"))
```


